---
title: "DGE analysis of Astrocyte and Microglia across different cerebral regions"
author: "Link√∂ping University, Joseph Filip Agi Maqdissi"
date: "`r Sys.Date()`"
output:
  pdf_document: default
  html_document: default
---

This is a protocol for the analysis of the data preformed in R and the packages being used in Rstudio. Read through the code and comments carefully. The protocole displayes the output in the terminal and the plots generated by the code. 

If there is any questions feel free to mail: Josag897@student.liu.se

```{r}

# loading necessary library
library(pathview)
library(clusterProfiler)
library(dplyr)
library(limma)
library(edgeR)
library(org.Mm.eg.db)
library(AnnotationDbi)
library(tidyverse)
library(biomaRt)
library(readxl)
library(affy)
library(ggplot2)
library(sva)
library(gplots)
library(VennDiagram)
library(eulerr)
library(UpSetR)
library(pheatmap)
library(writexl)
#  Load file with Counts per million after normalization data
rawCounts <-
  read.table("raw_counts_cpm_after_norm_reorganised_220722.txt", header = T)

# Converting ensemblID to symbol
rawCounts$GeneNames <- mapIds(
  org.Mm.eg.db,
  keys = rawCounts$ensembl_id,
  keytype = "ENSEMBL",
  column = "SYMBOL"
)


# Removing NA values
rawCounts <- na.omit(rawCounts)

# Keep only rows where all values are above 1
rawCounts <- rawCounts[apply(rawCounts, 1, function(x) all(x > 1)), ]


# Removing ENsemblID column
rawCounts <- rawCounts[, -1]
# Order the columns
rawCounts <- rawCounts[order(colnames(rawCounts))]

# Build a separated dataframe for the ribotag and totalRNA for astrocyte and microglia sep.
AC_RT <-
  subset(rawCounts,
         !(rownames(rawCounts) %in% "16709"),
         select = c(1, 2, 6, 7, 8, 12, 13, 14, 18))



MG_RT <-
  subset(rawCounts,
         !(rownames(rawCounts) %in% "16709"),
         select = c(19, 20, 21, 25, 26, 27, 31, 32, 33, 18))


# Check the variance of the gene expression data
AC_RT_var <- sapply(AC_RT, var)

MG_RT_var <- sapply(MG_RT, var)

# Plot the variance distribution to check if it is homogeneous
hist(AC_RT_var,
     main = "Histogram of Ribotag Astrocyte Values Distribution",
     xlab = "Values",
     col = "lightblue") # plot histogram

hist(MG_RT_var,
     main = "Histogram of Ribotag Microglia Values Distribution",
     xlab = "Values",
     col = "lightblue") # plot histogram


# Make object to  store p-values
PvalueAC_FM <- data.frame(Pvalue = rep(0, nrow(AC_RT)))
PvalueAC_FR <- data.frame(Pvalue = rep(0, nrow(AC_RT)))
PvalueAC_MR <- data.frame(Pvalue = rep(0, nrow(AC_RT)))

PvalueMG_FM <- data.frame(Pvalue = rep(0, nrow(MG_RT)))
PvalueMG_FR <- data.frame(Pvalue = rep(0, nrow(MG_RT)))
PvalueMG_MR <- data.frame(Pvalue = rep(0, nrow(MG_RT)))


for (i in 1:nrow(AC_RT)) {
  PvalueAC_FM$Pvalue[i] <-
    t.test(AC_RT[i, 1:2], AC_RT[i, 3:5], alternative = "two.sided")$p.value
  PvalueAC_FR$Pvalue[i] <-
    t.test(AC_RT[i, 1:2], AC_RT[i, 6:8], alternative = "two.sided")$p.value
  PvalueAC_MR$Pvalue[i] <-
    t.test(AC_RT[i, 3:5], AC_RT[i, 6:8], alternative = "two.sided")$p.value
}

for (i in 1:nrow(MG_RT)) {
  PvalueMG_FM$Pvalue[i] <-
    t.test(MG_RT[i, 1:3], MG_RT[i, 4:6], alternative = "two.sided")$p.value
  PvalueMG_FR$Pvalue[i] <-
    t.test(MG_RT[i, 1:3], MG_RT[i, 7:9], alternative = "two.sided")$p.value
  PvalueMG_MR$Pvalue[i] <-
    t.test(MG_RT[i, 4:6], MG_RT[i, 7:9], alternative = "two.sided")$p.value
}

# add the column genename
PvalueAC_FM$GeneNames <- AC_RT$GeneNames
PvalueAC_FR$GeneNames <- AC_RT$GeneNames
PvalueAC_MR$GeneNames <- AC_RT$GeneNames

PvalueMG_FM$GeneNames <- MG_RT$GeneNames
PvalueMG_FR$GeneNames <- MG_RT$GeneNames
PvalueMG_MR$GeneNames <- MG_RT$GeneNames


# make a FClog column into same pValobject and filter based onm p-value
PvalueAC_FM$log2FC <-
  rowMeans(AC_RT[, 1:2]) - rowMeans(AC_RT[, 3:5])
PvalueAC_FR$log2FC <-
  rowMeans(AC_RT[, 1:2]) - rowMeans(AC_RT[, 6:8])
PvalueAC_MR$log2FC <-
  rowMeans(AC_RT[, 3:5]) - rowMeans(AC_RT[, 6:8])

PvalueMG_FM$log2FC <-
  rowMeans(MG_RT[, 1:3]) - rowMeans(MG_RT[, 4:6])
PvalueMG_FR$log2FC <-
  rowMeans(MG_RT[, 1:3]) - rowMeans(MG_RT[, 7:9])
PvalueMG_MR$log2FC <-
  rowMeans(MG_RT[, 4:6]) - rowMeans(MG_RT[, 7:9])


#Removing the P-values over 0.05 for all p-values objects
#AC
sign_PvalueAC_FM <-
  PvalueAC_FM[PvalueAC_FM$Pvalue < 0.05, , drop = F]
sign_PvalueAC_FM <- na.omit(sign_PvalueAC_FM)

sign_PvalueAC_FR <-
  PvalueAC_FR[PvalueAC_FR$Pvalue < 0.05, , drop = F]
sign_PvalueAC_FR <- na.omit(sign_PvalueAC_FR)

sign_PvalueAC_MR <-
  PvalueAC_MR[PvalueAC_MR$Pvalue < 0.05, , drop = F]
sign_PvalueAC_MR <- na.omit(sign_PvalueAC_MR)

# MG
sign_PvalueMG_FM <-
  PvalueMG_FM[PvalueMG_FM$Pvalue < 0.05, , drop = F]
sign_PvalueMG_FM <- na.omit(sign_PvalueMG_FM)

sign_PvalueMG_FR <-
  PvalueMG_FR[PvalueMG_FR$Pvalue < 0.05, , drop = F]
sign_PvalueMG_FR <- na.omit(sign_PvalueMG_FR)

sign_PvalueMG_MR <-
  PvalueMG_MR[PvalueMG_MR$Pvalue < 0.05, , drop = F]
sign_PvalueMG_MR <- na.omit(sign_PvalueMG_MR)




# Convert gene symbols to ENTREZIDs
Deg_AC_FM <-
  bitr(
    sign_PvalueAC_FM$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )
Deg_AC_FR <-
  bitr(
    sign_PvalueAC_FR$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )
Deg_AC_MR <-
  bitr(
    sign_PvalueAC_MR$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )

Deg_MG_FM <-
  bitr(
    sign_PvalueMG_FM$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )
Deg_MG_FR <-
  bitr(
    sign_PvalueMG_FR$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )
Deg_MG_MR <-
  bitr(
    sign_PvalueMG_MR$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )

# One gene was converted twice by bitr in the sign_PvslueMG_FR object, thus it contains duplicates
# Remove one of the duplicates
sign_PvalueMG_FR <-
  sign_PvalueMG_FR[!duplicated(sign_PvalueMG_FR$GeneNames),]
# Same
sign_PvalueAC_MR <-
  sign_PvalueAC_MR[!duplicated(sign_PvalueAC_MR$GeneNames),]

##
# TOO do
# Pathway analysis


# create a background gene file on all of the genes
Backround_Genes <-
  data.frame(rawCounts$GeneNames) # create a data frame with the names of all measured genes

Backround_Genes <-
  bitr(
    Backround_Genes$rawCounts.GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  ) # convert symbol to ENTREZID

table(duplicated(Backround_Genes$ENTREZID)) #check if duplicated entrez ID


# AC
keggAstrocyte_FM <- enrichKEGG(
  gene = Deg_AC_FM$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggAstrocyte_FM <-
  setReadable(keggAstrocyte_FM, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggAstrocyte_summary_FM <- keggAstrocyte_FM@result
keggAstrocyte_summary_FM <-
  keggAstrocyte_summary_FM[keggAstrocyte_summary_FM$pvalue < 0.05, , drop = F]

#
keggAstrocyte_FR <- enrichKEGG(
  gene = Deg_AC_FR$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggAstrocyte_FM <-
  setReadable(keggAstrocyte_FR, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggAstrocyte_summary_FR <- keggAstrocyte_FR@result
keggAstrocyte_summary_FR <-
  keggAstrocyte_summary_FR[keggAstrocyte_summary_FR$pvalue < 0.05, , drop = F]

#
keggAstrocyte_MR <- enrichKEGG(
  gene = Deg_AC_MR$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggAstrocyte_MR <-
  setReadable(keggAstrocyte_MR, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggAstrocyte_summary_MR <- keggAstrocyte_MR@result
keggAstrocyte_summary_MR <-
  keggAstrocyte_summary_MR[keggAstrocyte_summary_MR$pvalue < 0.05, , drop = F]

# KEGG pathway analysis
# MG
keggMicroglia_FM <- enrichKEGG(
  gene = Deg_MG_FM$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggMicroglia_FM <-
  setReadable(keggMicroglia_FM, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggMicroglia_summary_FM <- keggMicroglia_FM@result
keggMicroglia_summary_FM <-
  keggMicroglia_summary_FM[keggMicroglia_summary_FM$pvalue < 0.05, , drop = F]

#
keggMicroglia_FR <- enrichKEGG(
  gene = Deg_MG_FR$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggMicroglia_FM <-
  setReadable(keggMicroglia_FR, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggMicroglia_summary_FR <- keggMicroglia_FR@result
keggMicroglia_summary_FR <-
  keggMicroglia_summary_FR[keggMicroglia_summary_FR$pvalue < 0.05, , drop = F]

#
keggMicroglia_MR <- enrichKEGG(
  gene = Deg_MG_MR$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggMicroglia_MR <-
  setReadable(keggMicroglia_MR, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggMicroglia_summary_MR <- keggMicroglia_MR@result
keggMicroglia_summary_MR <-
  keggMicroglia_summary_MR[keggMicroglia_summary_MR$pvalue < 0.05, , drop = F]
########################

create_custom_dot_plot <- function(Kegg_result) {
  # Create a new column 'geneRatioDecimal' in our dataframe which holds the numeric ratio
  Kegg_result$GeneRatioDecimal <-
    as.numeric(unlist(lapply(strsplit(as.character(Kegg_result$GeneRatio), "/"), function(x)
      as.numeric(x[1]) / as.numeric(x[2]))))
  
  dot_plot <-
    ggplot(Kegg_result,
           aes(
             x = reorder(Description,-GeneRatioDecimal),
             y = GeneRatioDecimal,
             color = pvalue,
             size = Count
           )) +
    geom_point() +
    scale_color_gradient(low = "blue", high = "red") +
    theme_minimal() +
    theme(
      axis.title.x = element_blank(),
      axis.text.x = element_text(
        size = 8,
        angle = 75,
        hjust = 1
      ),
      plot.title = element_text(size = 12, hjust = 0.5),
      panel.background = element_rect(fill = "gray90"),
      plot.background = element_rect(fill = "white")
    ) +
    labs(title = "Kegg Enrichment plot",
         y = "GeneRatio",
         x = "Description")
  
  
  return(dot_plot)
}

dot_plot_AC_FM <- create_custom_dot_plot(keggAstrocyte_summary_FM)
dot_plot_AC_FR <- create_custom_dot_plot(keggAstrocyte_summary_FR)
dot_plot_AC_MR <- create_custom_dot_plot(keggAstrocyte_summary_MR)

dot_plot_MG_FM <- create_custom_dot_plot(keggMicroglia_summary_FM)
dot_plot_MG_FR <- create_custom_dot_plot(keggMicroglia_summary_FR)
dot_plot_MG_MR <- create_custom_dot_plot(keggMicroglia_summary_MR)



#AC
log2fc_AC_FM <- data.frame(sign_PvalueAC_FM$log2FC)
rownames(log2fc_AC_FM) <- Deg_AC_FM$ENTREZID
log2fc_AC_FM <- na.omit(log2fc_AC_FM)

log2fc_AC_FR <- data.frame(sign_PvalueAC_FR$log2FC)
rownames(log2fc_AC_FR) <- Deg_AC_FR$ENTREZID
log2fc_AC_FR <- na.omit(log2fc_AC_FR)


log2fc_AC_MR <- data.frame(sign_PvalueAC_MR$log2FC)
rownames(log2fc_AC_MR) <- Deg_AC_MR$ENTREZID
log2fc_AC_MR <- na.omit(log2fc_AC_MR)


# MG
log2fc_MG_FM <- data.frame(sign_PvalueMG_FM$log2FC)
rownames(log2fc_MG_FM) <- Deg_MG_FM$ENTREZID
log2fc_MG_FM <- na.omit(log2fc_MG_FM)


log2fc_MG_FR <- data.frame(sign_PvalueMG_FR$log2FC)
rownames(log2fc_MG_FR) <- Deg_MG_FR$ENTREZID
log2fc_MG_FR <- na.omit(log2fc_MG_FR)


log2fc_MG_MR <- data.frame(sign_PvalueMG_MR$log2FC)
rownames(log2fc_MG_MR) <- Deg_MG_MR$ENTREZID
log2fc_MG_MR <- na.omit(log2fc_MG_MR)

# Pathview analysis
# AC_FM
pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu05033",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu04080",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)



pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu04659",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu05208",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu04060",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

# MG_FM
pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu05031",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu04728",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu05030",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu04024",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu05033",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu04024",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

# AC_FR
pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu04080",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu04934",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu04727",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu00900",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu00562",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)


# MG_FR
pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu04724",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu04020",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu04072",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu05031",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu04713",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu04080",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)



# AC_MR
pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu04080",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu04724",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu05033",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)



pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu04727",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)


# MG_MR
pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu04020",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu04970",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu04724",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu04024",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu04925",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu05414",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)



# AC
num_log2fc_AC_FM <- sign_PvalueAC_FM$log2FC
names(num_log2fc_AC_FM) <- Deg_AC_FM$ENTREZID

num_log2fc_AC_FR <- sign_PvalueAC_FR$log2FC
names(num_log2fc_AC_FR) <- Deg_AC_FR$ENTREZID

num_log2fc_AC_MR <- sign_PvalueAC_MR$log2FC
names(num_log2fc_AC_MR) <- Deg_AC_MR$ENTREZID


# MG
num_log2fc_MG_FM <- sign_PvalueMG_FM$log2FC
names(num_log2fc_MG_FM) <- Deg_MG_FM$ENTREZID

num_log2fc_MG_FR <- sign_PvalueMG_FR$log2FC
names(num_log2fc_MG_FR) <- Deg_MG_FR$ENTREZID

num_log2fc_MG_MR <- sign_PvalueMG_MR$log2FC
names(num_log2fc_MG_MR) <- Deg_MG_MR$ENTREZID

# AC
Sort_log2fc_AC_FM <- sort(num_log2fc_AC_FM, decreasing = TRUE)
Sort_log2fc_AC_FR <- sort(num_log2fc_AC_FR, decreasing = TRUE)
Sort_log2fc_AC_MR <- sort(num_log2fc_AC_MR, decreasing = TRUE)

# MG
Sort_log2fc_MG_FM <- sort(num_log2fc_MG_FM, decreasing = TRUE)
Sort_log2fc_MG_FR <- sort(num_log2fc_MG_FR, decreasing = TRUE)
Sort_log2fc_MG_MR <- sort(num_log2fc_MG_MR, decreasing = TRUE)



# AC
gseaKEGG_AC_FM <- gseKEGG(
  geneList = Sort_log2fc_AC_FM,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus
gseaKEGG_AC_FR <- gseKEGG(
  geneList = Sort_log2fc_AC_FR,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus
gseaKEGG_AC_MR <- gseKEGG(
  geneList = Sort_log2fc_AC_MR,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus

# MG
gseaKEGG_MG_FM <- gseKEGG(
  geneList = Sort_log2fc_MG_FM,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus
gseaKEGG_MG_FR <- gseKEGG(
  geneList = Sort_log2fc_MG_FR,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus
gseaKEGG_MG_MR <- gseKEGG(
  geneList = Sort_log2fc_MG_MR,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus


View(gseaKEGG_AC_FM@result)
View(gseaKEGG_AC_FR@result)
View(gseaKEGG_AC_MR@result)

View(gseaKEGG_MG_FM@result)
View(gseaKEGG_MG_FR@result)
View(gseaKEGG_MG_MR@result)

write_xlsx(gseaKEGG_AC_FM@result, "GSEA_AC_FM.xlsx")
write_xlsx(gseaKEGG_AC_FR@result, "GSEA_AC_FR.xlsx")
write_xlsx(gseaKEGG_AC_MR@result, "GSEA_AC_MR.xlsx")

write_xlsx(gseaKEGG_MG_FM@result, "GSEA_MG_FM.xlsx")
write_xlsx(gseaKEGG_MG_FR@result, "GSEA_MG_FR.xlsx")
write_xlsx(gseaKEGG_MG_MR@result, "GSEA_MG_MR.xlsx")

# Plot the GSEA plot for one of the enriched pathways observed in the enrichedKEGG object
# Gen Set ID mmu00010
# AC
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05415", title = "GSEA enrichment score plot Astrocytes Front & Middle - Diabetic cardiomyopathy")

gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05010", title = "GSEA enrichment score plot Astrocytes Front & Middle - Alzheimer disease")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu01100", title = "GSEA enrichment score plot Astrocytes Front & Middle - Metabolic pathways")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05020", title = "GSEA enrichment score plot Astrocytes Front & Middle - Prion disease")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05208", title = "GSEA enrichment score plot Astrocytes Front & Middle - Chemical carcinogenesis")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05012", title = "GSEA enrichment score plot Astrocytes Front & Middle - Parkinson disease")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu04714", title = "GSEA enrichment score plot Astrocytes Front & Middle - Thermogenesis")


gseaplot(gseaKEGG_AC_FR, geneSetID = "mmu04721", title = "GSEA enrichment score plot Astrocytes Front & Rare - Synaptic vesicle cycle")
gseaplot(gseaKEGG_AC_FR, geneSetID = "mmu01230", title = "GSEA enrichment score plot Astrocytes Front & Rare - Biosynthesis of amino acids")
gseaplot(gseaKEGG_AC_FR, geneSetID = "mmu04210", title = "GSEA enrichment score plot Astrocytes Front & Rare - Apoptosis")


gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu03010", title = "GSEA enrichment score plot Astrocytes Middle & Rare - Ribosome")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu00250", title = "GSEA enrichment score plot Astrocytes Middle & Rare - Alanine, aspartate and glutamate metabolism")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu04217", title = "GSEA enrichment score plot Astrocytes Middle & Rare - Necroptosis")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu04210", title = "GSEA enrichment score plot Astrocytes Middle & Rare - Apoptosis")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu05223", title = "GSEA enrichment score plot Astrocytes Middle & Rare - Non-small cell lung cancer")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu01230", title = "GSEA enrichment score plot Astrocytes Middle & Rare - Biosynthesis of amino acids")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu01040", title = "GSEA enrichment score plot Astrocytes Middle & Rare - Biosynthesis of unsaturated fatty acids")


# MG
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu04020", title = "GSEA enrichment score plot Microglia Front & Middle - Calcium signaling pathway")
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu04728", title = "GSEA enrichment score plot Microglia Front & Middle - Dopaminergic synapse")
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu04024", title = "GSEA enrichment score plot Microglia Front & Middle - cAMP signaling pathway")
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu01100", title = "GSEA enrichment score plot Microglia Front & Middle - Metabolic pathways")
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu05016", title = "GSEA enrichment score plot Microglia Front & Middle - Huntington disease")


gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu01100", title = "GSEA enrichment score plot Microglia Front & Rare - Metabolic pathways")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu05417", title = "GSEA enrichment score plot Microglia Front & Rare - Lipid and atherosclerosis")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu04970", title = "GSEA enrichment score plot Microglia Front & Rare - Salivary secretion")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu01200", title = "GSEA enrichment score plot Microglia Front & Rare - Carbon metabolism")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu04970", title = "GSEA enrichment score plot Microglia Front & Rare - Salivary secretion")


gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu01100", title = "GSEA enrichment score plot Microglia Middle & Rare - Metabolic pathways")
gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu04971", title = "GSEA enrichment score plot Microglia Middle & Rare - Gastric acid secretion")
gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu04114", title = "GSEA enrichment score plot Microglia Middle & Rare - Oocyte meiosis")
gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu05417", title = "GSEA enrichment score plot Microglia Middle & Rare - Lipid and atherosclerosis")
gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu04310", title = "GSEA enrichment score plot Microglia Middle & Rare - Wnt signaling pathway")


# Function to create and save a plot
save_gseaplot <- function(gsea_result, geneSetID, title, file_name) {
  plot <- gseaplot(gsea_result, geneSetID = geneSetID, title = title)
  ggsave(file_name, plot, width = 15, height = 8.5, dpi = 300)
}

# AC
save_gseaplot(gseaKEGG_AC_FM, "mmu05415", "GSEA enrichment score plot Astrocytes Front & Middle - Diabetic cardiomyopathy", "GSEA_AC_FM_Diabetic_cardiomyopathy.png")
save_gseaplot(gseaKEGG_AC_FM, "mmu05010", "GSEA enrichment score plot Astrocytes Front & Middle - Alzheimer disease", "GSEA_AC_FM_Alzheimer_disease.png")
save_gseaplot(gseaKEGG_AC_FM, "mmu01100", "GSEA enrichment score plot Astrocytes Front & Middle - Metabolic pathways", "GSEA_AC_FM_Metabolic_pathways.png")
save_gseaplot(gseaKEGG_AC_FM, "mmu05020", "GSEA enrichment score plot Astrocytes Front & Middle - Prion disease", "GSEA_AC_FM_Prion_disease.png")
save_gseaplot(gseaKEGG_AC_FM, "mmu05208", "GSEA enrichment score plot Astrocytes Front & Middle - Chemical carcinogenesis", "GSEA_AC_FM_Chemical_carcinogenesis.png")
save_gseaplot(gseaKEGG_AC_FM, "mmu05012", "GSEA enrichment score plot Astrocytes Front & Middle - Parkinson disease", "GSEA_AC_FM_Parkinson_disease.png")
save_gseaplot(gseaKEGG_AC_FM, "mmu04714", "GSEA enrichment score plot Astrocytes Front & Middle - Thermogenesis", "GSEA_AC_FM_Thermogenesis.png")

save_gseaplot(gseaKEGG_AC_FR, "mmu04721", "GSEA enrichment score plot Astrocytes Front & Rare - Synaptic vesicle cycle", "GSEA_AC_FR_Synaptic_vesicle_cycle.png")
save_gseaplot(gseaKEGG_AC_FR, "mmu01230", "GSEA enrichment score plot Astrocytes Front & Rare - Biosynthesis of amino acids", "GSEA_AC_FR_Biosynthesis_of_amino_acids.png")
save_gseaplot(gseaKEGG_AC_FR, "mmu04210", "GSEA enrichment score plot Astrocytes Front & Rare - Apoptosis", "GSEA_AC_FR_Apoptosis.png")

save_gseaplot(gseaKEGG_AC_MR, "mmu03010", "GSEA enrichment score plot Astrocytes Middle & Rare - Ribosome", "GSEA_AC_MR_Ribosome.png")
save_gseaplot(gseaKEGG_AC_MR, "mmu00250", "GSEA enrichment score plot Astrocytes Middle & Rare - Alanine, aspartate and glutamate metabolism", "GSEA_AC_MR_Alanine_aspartate_glutamate_metabolism.png")
save_gseaplot(gseaKEGG_AC_MR, "mmu04217", "GSEA enrichment score plot Astrocytes Middle & Rare - Necroptosis", "GSEA_AC_MR_Necroptosis.png")
save_gseaplot(gseaKEGG_AC_MR, "mmu04210", "GSEA enrichment score plot Astrocytes Middle & Rare - Apoptosis", "GSEA_AC_MR_Apoptosis.png")
save_gseaplot(gseaKEGG_AC_MR, "mmu05223", "GSEA enrichment score plot Astrocytes Middle & Rare - Non-small cell lung cancer", "GSEA_AC_MR_Non-small_cell_lung_cancer.png")
save_gseaplot(gseaKEGG_AC_MR, "mmu01230", "GSEA enrichment score plot Astrocytes Middle & Rare - Biosynthesis of amino acids", "GSEA_AC_MR_Biosynthesis_of_amino_acids.png")
save_gseaplot(gseaKEGG_AC_MR, "mmu01040", "GSEA enrichment score plot Astrocytes Middle & Rare - Biosynthesis of unsaturated fatty acids", "GSEA_AC_MR_Biosynthesis_of_unsaturated_fatty_acids.png")

# MG
save_gseaplot(gseaKEGG_MG_FM, "mmu04020", "GSEA enrichment score plot Microglia Front & Middle - Calcium signaling pathway", "GSEA_MG_FM_Calcium_signaling_pathway.png")
save_gseaplot(gseaKEGG_MG_FM, "mmu04728", "GSEA enrichment score plot Microglia Front & Middle - Dopaminergic synapse", "GSEA_MG_FM_Dopaminergic_synapse.png")
save_gseaplot(gseaKEGG_MG_FM, "mmu04024", "GSEA enrichment score plot Microglia Front & Middle - cAMP signaling pathway", "GSEA_MG_FM_cAMP_signaling_pathway.png")
save_gseaplot(gseaKEGG_MG_FM, "mmu01100", "GSEA enrichment score plot Microglia Front & Middle - Metabolic pathways", "GSEA_MG_FM_Metabolic_pathways.png")
save_gseaplot(gseaKEGG_MG_FM, "mmu05016", "GSEA enrichment score plot Microglia Front & Middle - Huntington disease", "GSEA_MG_FM_Huntington_disease.png")

save_gseaplot(gseaKEGG_MG_FR, "mmu01100", "GSEA enrichment score plot Microglia Front & Rare - Metabolic pathways", "GSEA_MG_FR_Metabolic_pathways.png")
save_gseaplot(gseaKEGG_MG_FR, "mmu05417", "GSEA enrichment score plot Microglia Front & Rare - Lipid and atherosclerosis", "GSEA_MG_FR_Lipid_and_atherosclerosis.png")
save_gseaplot(gseaKEGG_MG_FR, "mmu04970", "GSEA enrichment score plot Microglia Front & Rare - Salivary secretion", "GSEA_MG_FR_Salivary_secretion.png")
save_gseaplot(gseaKEGG_MG_FR, "mmu01200", "GSEA enrichment score plot Microglia Front & Rare - Carbon metabolism", "GSEA_MG_FR_Carbon_metabolism.png")
save_gseaplot(gseaKEGG_MG_FR, "mmu04970", "GSEA enrichment score plot Microglia Front & Rare - Salivary secretion", "GSEA_MG_FR_Salivary_secretion.png")

save_gseaplot(gseaKEGG_MG_MR, "mmu01100", "GSEA enrichment score plot Microglia Middle & Rare - Metabolic pathways", "GSEA_MG_MR_Metabolic_pathways.png")
save_gseaplot(gseaKEGG_MG_MR, "mmu04971", "GSEA enrichment score plot Microglia Middle & Rare - Gastric acid secretion", "GSEA_MG_MR_Gastric_acid_secretion.png")
save_gseaplot(gseaKEGG_MG_MR, "mmu04114", "GSEA enrichment score plot Microglia Middle & Rare - Oocyte meiosis", "GSEA_MG_MR_Oocyte_meiosis.png")
save_gseaplot(gseaKEGG_MG_MR, "mmu05417", "GSEA enrichment score plot Microglia Middle & Rare - Lipid and atherosclerosis", "GSEA_MG_MR_Lipid_and_atherosclerosis.png")
save_gseaplot(gseaKEGG_MG_MR, "mmu04310", "GSEA enrichment score plot Microglia Middle & Rare - Wnt signaling pathway", "GSEA_MG_MR_Wnt_signaling_pathway.png")


## Save dotplots from KEGG enrichment plot
ggsave(
  filename = "dot_plot_AC_FM.png",
  plot = dot_plot_AC_FM,
  width = 10,
  height = 12,
  dpi = 300
)
ggsave(
  filename = "dot_plot_AC_FR.png",
  plot = dot_plot_AC_FR,
  width = 10,
  height = 12,
  dpi = 300
)
ggsave(
  filename = "dot_plot_AC_MR.png",
  plot = dot_plot_AC_MR,
  width = 10,
  height = 12,
  dpi = 300
)

ggsave(
  filename = "dot_plot_MG_FM.png",
  plot = dot_plot_MG_FM,
  width = 10,
  height = 12,
  dpi = 300
)
ggsave(
  filename = "dot_plot_MG_FR.png",
  plot = dot_plot_MG_FR,
  width = 10,
  height = 12,
  dpi = 300
)
ggsave(
  filename = "dot_plot_MG_MR.png",
  plot = dot_plot_MG_MR,
  width = 10,
  height = 12,
  dpi = 300
)




```

```{r}
# read the data file with all samples
rawCounts <-
  read.table("raw_counts_cpm_after_norm_reorganised_220722.txt", header = TRUE)


# Load sample information data
sampleInfo <- read.table("Sample-info_220722.txt", header = TRUE)


# Load cell type marker genes data from an Excel file
cellType_markerGenes <- read_excel("Celltype_markergenes.xlsx")



# USing biomaRt annotation  method.

# WARNING THIS FUNCTION DOES NOT WORK we USE OLD LOADED DATASET FOR GENESYMBOLS Y THIS FUNCTION 2023-12-27.
# ensembl <- useMart("ensembl")
# 
# listDatasets(ensembl)
# 
# ensembl_mouse <- useDataset("mmusculus_gene_ensembl", mart = ensembl)
# 
# mouse_gene_ids <- rawCounts$ensembl_id
# THIS FUNCTION getBM() SHOULD WORK BUT HAS A TIME TOKEN LIMIT ERROR ON THE collect() function THAT NEEDS DEBUGGING.
# mouse_gene_info <-
#   getBM(
#     attributes = c(
#       "ensembl_gene_id",
#       "external_gene_name",
#       "chromosome_name",
#       "start_position",
#       "end_position",
#       "gene_biotype",
#       "strand"
#     ),
#     filters = "ensembl_gene_id",
#     values = mouse_gene_ids,
#     mart = ensembl_mouse
#   )
mouse_gene_info <- readRDS("mouse_gene_info.rds")
# Changing name of column too merge the values with a matching column.
colnames(mouse_gene_info)[c(1)] <- c("ensembl_id")
# Merging the two dataframes into one dataframe leaving out one
rawCounts <- merge(rawCounts, mouse_gene_info, c(1))

rawCounts <- rawCounts[,-c(1, 38:42)]

colnames(rawCounts)[c(36)] <-  c("GeneNames")

# keep only the first occurrence of each duplicated gene name
rawCounts <- rawCounts[!duplicated(rawCounts$GeneNames),]

# add gene name to  row names 
row.names(rawCounts) <-  rawCounts$GeneNames
rawCounts <- rawCounts[,-36]

# Keep only rows where all values are above 1
#rawCounts <- rawCounts[apply(rawCounts, 1, function(x) all(x > 1)), ]


# Filter the data using logical indexing
# rawCounts <- rawCounts[rowSums(rawCounts >= 40) > 0, ]

# separating the columns by region
# into  separated dataframe object for the rawCounts and sampleInfo
rawCounts_F <- rawCounts[, grep("_F_", colnames(rawCounts))]
rawCounts_M <- rawCounts[, grep("_M_", colnames(rawCounts))]
rawCounts_R <- rawCounts[, grep("_R_", colnames(rawCounts))]


sampleInfo_F <- sampleInfo[grep("_F_", sampleInfo$sample),]
sampleInfo_M <- sampleInfo[grep("_M_", sampleInfo$sample),]
sampleInfo_R <- sampleInfo[grep("_R_", sampleInfo$sample),]



# order Column names 
rawCounts <- rawCounts[order(colnames(rawCounts))]
sampleInfo <- sampleInfo[order(sampleInfo$sample), ]

rawCounts_F <- rawCounts_F[order(colnames(rawCounts_F))]
rawCounts_M <- rawCounts_M[order(colnames(rawCounts_M))]

sampleInfo_F <- sampleInfo_F[order(sampleInfo_F$sample), ]
sampleInfo_M <- sampleInfo_M[order(sampleInfo_M$sample), ]

rawCounts_R <- rawCounts_R[order(colnames(rawCounts_R))]
sampleInfo_R <- sampleInfo_R[order(sampleInfo_R$sample), ]


# Log the counts
logCounts <- log2(rawCounts + 1)

logCounts_F <- log2(rawCounts_F + 1)
logCounts_M <- log2(rawCounts_M + 1)
logCounts_R <- log2(rawCounts_R + 1)

###

# Adding marker genes before z-score
dfCounts <- na.omit(logCounts[cellType_markerGenes$gene_symbol, ])
dfCounts_F <- na.omit(logCounts_F[cellType_markerGenes$gene_symbol, ])
dfCounts_M <- na.omit(logCounts_M[cellType_markerGenes$gene_symbol, ])
dfCounts_R <- na.omit(logCounts_R[cellType_markerGenes$gene_symbol, ])

# making function for calculating z score since scale function is not ideal
calc_zscore <- function(df) {
  result <- (df - mean(df)) / sd(df)
  return(result)
}


# calculating the z-score
Z_dfCounts <- apply(dfCounts, 1, calc_zscore)
Z_dfCounts_F <- apply(dfCounts_F, 1, calc_zscore)
Z_dfCounts_M <- apply(dfCounts_M, 1, calc_zscore)
Z_dfCounts_R <- apply(dfCounts_R, 1, calc_zscore)


# Transposition original data frame
Z_dfCounts <- t(Z_dfCounts)
Z_dfCounts_F <- t(Z_dfCounts_F)
Z_dfCounts_M <- t(Z_dfCounts_M)
Z_dfCounts_R <- t(Z_dfCounts_R)

# Separate the columns into RT and TR
RT_columns <- grep("_RT", colnames(Z_dfCounts), value = TRUE)
TR_columns <- grep("_TR", colnames(Z_dfCounts), value = TRUE)

RT_columns_F <- grep("_RT", colnames(Z_dfCounts_F), value = TRUE)
TR_columns_F <- grep("_TR", colnames(Z_dfCounts_F), value = TRUE)

RT_columns_M <- grep("_RT", colnames(Z_dfCounts_M), value = TRUE)
TR_columns_M <- grep("_TR", colnames(Z_dfCounts_M), value = TRUE)


RT_columns_R <- grep("_RT", colnames(Z_dfCounts_R), value = TRUE)
TR_columns_R <- grep("_TR", colnames(Z_dfCounts_R), value = TRUE)


# Combine the columns in the desired order
ordered_columns <- c(RT_columns, TR_columns)
ordered_columns_F <- c(RT_columns_F, TR_columns_F)
ordered_columns_M <- c(RT_columns_M, TR_columns_M)
ordered_columns_R <- c(RT_columns_R, TR_columns_R)

# Reorder the data frame based on these columns
Z_dfCounts_ordered <- Z_dfCounts[, ordered_columns]
Z_dfCounts_F_ordered <-  Z_dfCounts_F[, ordered_columns_F]
Z_dfCounts_M_ordered <-  Z_dfCounts_M[, ordered_columns_M]
Z_dfCounts_R_ordered <-  Z_dfCounts_R[, ordered_columns_R]

# Create a new data frame with gene_symbol and cell_type for row annotation
row_annotation <- data.frame(
  CellType = cellType_markerGenes$cell_type,
  row.names = cellType_markerGenes$gene_symbol
)

col_annotation <- data.frame(Cells = sampleInfo$Cell_type,
                             Region = sampleInfo$Region,
                             Condition = sampleInfo$Condition,
row.names = sampleInfo$sample)

# Define a custom color scheme for cell types
cell_type_colors <- c("blue", "red", "green", "purple", "orange")
names(cell_type_colors) <- c("astro", "GABA", "Glut", "micro", "oligo")

cellType_colors <- c("#fbceb1", "#00ffff")
names(cellType_colors) <- c("Astrocytes", "Microglia")

region_colors <- c("#f0e442", "#d55e00", "#009e73")
names(region_colors) <- c("Front", "Middle", "Rear")

condition_color <- c("#FF00CC", "black")
names(condition_color) <- c("RiboTag", "TotalRNA")

# Generate the custom annotation dataframe
annotation_colors <- list(CellType = cell_type_colors,
                          Cells = cellType_colors,
                          Region = region_colors,
                          Condition = condition_color)
# Create the heatmap
FMR <- pheatmap(
  Z_dfCounts_ordered,
  color = colorRampPalette(c("yellow", "white", "purple"))(40),
  cluster_cols = F,
  cluster_rows = F,
  annotation_row = row_annotation,
  annotation_col = col_annotation,
  annotation_colors = annotation_colors,
  main = "Z-score of marker genes across cells Microglia & Astrocytes for all regions",
  fontsize = 8,
  show_colnames = F,
  cellwidth = 12,
  cellheight = 7
)

F1 <- pheatmap(
  Z_dfCounts_F_ordered,
  color = colorRampPalette(c("yellow", "white", "purple"))(40),
  cluster_cols = F,
  cluster_rows = F,
  annotation_row = row_annotation,
  annotation_col = col_annotation,
  annotation_colors = annotation_colors,
  main = "Z-score of marker genes across cells Microglia & Astrocytes for front region",
  fontsize = 8,
  cellwidth = 12,
  cellheight = 7
)


M1 <- pheatmap(
  Z_dfCounts_M_ordered,
  color = colorRampPalette(c("yellow", "white", "purple"))(40),
  cluster_cols = F,
  cluster_rows = F,
  annotation_row = row_annotation,
  annotation_col = col_annotation,
  annotation_colors = annotation_colors,
  main = "Z-score of marker genes across cells Microglia & Astrocytes for middle region",
  fontsize = 8,
  show_colnames = F,
  cellwidth = 12,
  cellheight = 7
)

R1 <- pheatmap(
  Z_dfCounts_R_ordered,
  color = colorRampPalette(c("yellow", "white", "purple"))(40),
  cluster_cols = F,
  cluster_rows = F,
  annotation_row = row_annotation,
  annotation_col = col_annotation,
  annotation_colors = annotation_colors,
  main = "Z-score of marker genes across cells Microglia & Astrocytes for rear region",
  fontsize = 8,
  show_colnames = F,
  cellwidth = 12,
  cellheight = 7
)

ggsave(
  filename = "3Z-score_HeatmapMarkerFrontMiddleRear.png",
  plot = FMR,
  width = 13,
  height = 13
)

ggsave(
  filename = "3Z-score_HeatmapMarkerFront.png",
  plot = F1,
  width = 13,
  height = 13
)
ggsave(
  filename = "3Z-score_HeatmapMarkerMiddle.png",
  plot = M1,
  width = 13,
  height = 13
)
ggsave(
  filename = "3Z-score_HeatmapMarkerRear.png",
  plot = R1,
  width = 13,
  height = 13
)

```

```{r}
# read the data file with all samples
rawCounts <-
  read.table("raw_counts_cpm_after_norm_reorganised_220722.txt", header = TRUE)


# Load sample information data
sampleInfo <- read.table("Sample-info_220722.txt", header = TRUE)

# Load cell type marker genes data from an Excel file
ribo_prot <- read_xlsx("ribo_prot.xlsx")

# BioMart is not working properly.
# ensembl <- useMart("ensembl")
# 
# listDatasets(ensembl)
# 
# ensembl_mouse <-
#   useDataset("mmusculus_gene_ensembl", mart = ensembl)
# 
# mouse_gene_ids <- rawCounts$ensembl_id

# THIS FUNCTION getBM() SHOULD WORK BUT HAS A TIME TOKEN LIMIT ERROR ON THE collect() function THAT NEEDS DEBUGGING.
# mouse_gene_info <-
#   getBM(
#     attributes = c(
#       "ensembl_gene_id",
#       "external_gene_name",
#       "chromosome_name",
#       "start_position",
#       "end_position",
#       "gene_biotype",
#       "strand"
#     ),
#     filters = "ensembl_gene_id",
#     values = mouse_gene_ids,
#     mart = ensembl_mouse
#   )

# Changing name of column too merge the values with a matching column.
colnames(mouse_gene_info)[c(1)] <- c("ensembl_id")
# Merging the two dataframes into one dataframe leaving out one
rawCounts <- merge(rawCounts, mouse_gene_info, c(1))

rawCounts <- rawCounts[, -c(1, 38:42)]

colnames(rawCounts)[c(36)] <-  c("GeneNames")

# keep only the first occurrence of each duplicated gene name
rawCounts <- rawCounts[!duplicated(rawCounts$GeneNames), ]

# add gene name to  row names
row.names(rawCounts) <-  rawCounts$GeneNames
rawCounts <- rawCounts[, -36]

# # Keep only rows where all values are above 1
# rawCounts <- rawCounts[apply(rawCounts, 1, function(x) all(x > 1)), ]

# Filter the data using logical indexing
# rawCounts <- rawCounts[rowSums(rawCounts >= 40) > 0, ]

# separating the columns if ribotag and totalRNA  
# into  separated dataframe object for the rawCounts and sampleInfo
rawCounts_F <- rawCounts[, grep("_F_", colnames(rawCounts))]
rawCounts_M <- rawCounts[, grep("_M_", colnames(rawCounts))]
rawCounts_R <- rawCounts[, grep("_R_", colnames(rawCounts))]


sampleInfo_F <- sampleInfo[grep("_F_", sampleInfo$sample),]
sampleInfo_M <- sampleInfo[grep("_M_", sampleInfo$sample),]
sampleInfo_R <- sampleInfo[grep("_R_", sampleInfo$sample),]

# order Column names 
rawCounts <- rawCounts[order(colnames(rawCounts))]
sampleInfo <- sampleInfo[order(sampleInfo$sample), ]

rawCounts_F <- rawCounts_F[order(colnames(rawCounts_F))]
rawCounts_M <- rawCounts_M[order(colnames(rawCounts_M))]

sampleInfo_F <- sampleInfo_F[order(sampleInfo_F$sample), ]
sampleInfo_M <- sampleInfo_M[order(sampleInfo_M$sample), ]

rawCounts_R <- rawCounts_R[order(colnames(rawCounts_R))]
sampleInfo_R <- sampleInfo_R[order(sampleInfo_R$sample), ]

# Log the counts
logCounts <- log2(rawCounts + 1)

logCounts_F <- log2(rawCounts_F + 1)
logCounts_M <- log2(rawCounts_M + 1)
logCounts_R <- log2(rawCounts_R + 1)


# Adding marker genes before z-score
dfCounts_D <- na.omit(logCounts[ribo_prot$Symbols, ])
dfCounts_F_D <- na.omit(logCounts_F[ribo_prot$Symbols, ])
dfCounts_M_D <- na.omit(logCounts_M[ribo_prot$Symbols, ])
dfCounts_R_D <- na.omit(logCounts_R[ribo_prot$Symbols, ])

# making function for calculating z score since scale function is not ideal
calc_zscore <- function(df) {
  result <- (df - mean(df)) / sd(df)
  return(result)
}


# calculating the z-score
Z_dfCounts_D <- apply(dfCounts_D, 1, calc_zscore)
Z_dfCounts_F_D <- apply(dfCounts_F_D, 1, calc_zscore)
Z_dfCounts_M_D <- apply(dfCounts_M_D, 1, calc_zscore)
Z_dfCounts_R_D <- apply(dfCounts_R_D, 1, calc_zscore)



# # keep only the first occurrence of each duplicated gene name
# ribo_prot <- ribo_prot[!duplicated(ribo_prot$Symbols),]
# # Create a new data frame with gene_symbol and cell_type for row annotation
# row_annotation_D <- data.frame(
#   Cromosome = ribo_prot$`Chromosome Number`,
#   row.names = ribo_prot$Symbols
# )
# 
# # Define a custom color scheme for cell types
# 
# # Generate the custom annotation dataframe
# annotation_D <- data.frame(row_annotation_D)
# rownames(annotation_D) <- rownames(row_annotation_D)



# Your original data frame
Z_dfCounts_D <- t(Z_dfCounts_D)
Z_dfCounts_F_D <- t(Z_dfCounts_F_D)
Z_dfCounts_M_D <- t(Z_dfCounts_M_D)
Z_dfCounts_R_D <- t(Z_dfCounts_R_D)

# Separate the columns into RT and TR
RT_columns_D <- grep("_RT", colnames(Z_dfCounts_D), value = TRUE)
TR_columns_D <- grep("_TR", colnames(Z_dfCounts_D), value = TRUE)

RT_columns_F_D <- grep("_RT", colnames(Z_dfCounts_F_D), value = TRUE)
TR_columns_F_D <- grep("_TR", colnames(Z_dfCounts_F_D), value = TRUE)

RT_columns_M_D <- grep("_RT", colnames(Z_dfCounts_M_D), value = TRUE)
TR_columns_M_D <- grep("_TR", colnames(Z_dfCounts_M_D), value = TRUE)


RT_columns_R_D <- grep("_RT", colnames(Z_dfCounts_R_D), value = TRUE)
TR_columns_R_D <- grep("_TR", colnames(Z_dfCounts_R_D), value = TRUE)


# Combine the columns in the desired order
ordered_columns_D <- c(RT_columns_D, TR_columns_D)
ordered_columns_F_D <- c(RT_columns_F_D, TR_columns_F_D)
ordered_columns_M_D <- c(RT_columns_M_D, TR_columns_M_D)
ordered_columns_R_D <- c(RT_columns_R_D, TR_columns_R_D)

# Reorder the data frame based on these columns
Z_dfCounts_ordered_D <- Z_dfCounts_D[, ordered_columns_D]
Z_dfCounts_F_ordered_D <-  Z_dfCounts_F_D[, ordered_columns_F_D]
Z_dfCounts_M_ordered_D <-  Z_dfCounts_M_D[, ordered_columns_M_D]
Z_dfCounts_R_ordered_D <-  Z_dfCounts_R_D[, ordered_columns_R_D]


# Create the heatmap
FMR1 <- pheatmap(
  Z_dfCounts_ordered_D,
  color = colorRampPalette(c("yellow", "white", "purple"))(40),
  annotation_col = col_annotation,
  annotation_colors = annotation_colors,
  cluster_cols = F,
  cluster_rows = F,
  main = "Z-score of marker genes across cells Microglia & Astrocytes for all regions",
  fontsize = 8,
  show_colnames = F,
  cellwidth = 12,
  cellheight = 7
)

F2 <- pheatmap(
  Z_dfCounts_F_ordered_D,
  color = colorRampPalette(c("yellow", "white", "purple"))(40),
  annotation_col = col_annotation,
  annotation_colors = annotation_colors,
  cluster_cols = F,
  cluster_rows = F,
  main = "Z-score of ribosomal protein encoded genes across cells Microglia & Astrocytes for front region",
  fontsize = 8,
  show_colnames = F,
  cellwidth = 12,
  cellheight = 7
)


M2 <- pheatmap(
  Z_dfCounts_M_ordered_D,
  color = colorRampPalette(c("yellow", "white", "purple"))(40),
  annotation_col = col_annotation,
  annotation_colors = annotation_colors,
  show_colnames = F,
  cluster_cols = F,
  cluster_rows = F,
  main = "Z-score of ribosomal protein encoded genes across cells Microglia & Astrocytes for middle region",
  fontsize = 8,
  cellwidth = 12,
  cellheight = 7
)

R2 <- pheatmap(
  Z_dfCounts_R_ordered_D,
  color = colorRampPalette(c("yellow", "white", "purple"))(40),
  annotation_col = col_annotation,
  annotation_colors = annotation_colors,
  show_colnames = F,
  cluster_cols = F,
  cluster_rows = F,
  main = "Z-score of ribosomal protein encoded genes across cells Microglia & Astrocytes for rear region",
  fontsize = 8,
  cellwidth = 12,
  cellheight = 7
)

ggsave(
  filename = "3Z-score_HeatmapRiboProtFrontMiddleRear.png",
  plot = FMR1,
  width = 13,
  height = 13
)

ggsave(
  filename = "3Z-score_HeatmapRiboProtFront.png",
  plot = F2,
  width = 13,
  height = 13
)

ggsave(
  filename = "3Z-score_HeatmapRiboProtMiddle.png",
  plot = M2,
  width = 13,
  height = 13
)

ggsave(
  filename = "3Z-score_HeatmapRiboProtRear.png",
  plot = R2,
  width = 13,
  height = 13
)

```


```{r}





# Attaching data to object in the environment
# Sample Info
sampleInfo <- read.table("Sample-info_220722.txt", header = T)

# Celltype info
Celltype_markergenes <- read_excel("Celltype_markergenes.xlsx")

# Counts per million after normalization
rawCounts <-
  read.table(
    "raw_counts_cpm_after_norm_reorganised_220722.txt",
    header = T,
    row.names = 1
  )



# Set a seed to produce reproducible result
set.seed(16)

#Create DGE through DGEList() funtion
dgList <- DGEList(counts = rawCounts, genes = rownames(rawCounts))

# Add the replication column to the markergenes dataframe
# Building groups for the replication
replication <-
  factor(
    c(
      'rep1',
      'rep1',
      'rep1',
      'rep2',
      'rep2',
      'rep2',
      'rep3',
      'rep3',
      'rep2',
      'rep2',
      'rep2',
      'rep3',
      'rep3',
      'rep3',
      'rep1',
      'rep1',
      'rep1',
      'rep1',
      'rep1',
      'rep1',
      'rep2',
      'rep2',
      'rep2',
      'rep3',
      'rep3',
      'rep3',
      'rep2',
      'rep2',
      'rep2',
      'rep3',
      'rep3',
      'rep3',
      'rep1',
      'rep1',
      'rep1'
    )
  )

replication <- data.frame(replication)
sampleInfo$replication <- replication$replication



# Density plots of CPM per sample of data, filtered, and normalized log2
plotDensity(
  cpm(rawCounts, log = T),
  col = "red",
  xlab = "CPM of log2",
  ylab = "Density",
  main = "Normalized CPM log2 Data"
)

# Extract the Counts Matrix from the DGEList
Count_dgList <- dgList$counts

# Log the counts
logCounts_dgList <- log2(Count_dgList + 1)

# Filter out low-expressed genes
maxTPM <- apply(logCounts_dgList, 1, max)
minTPM <- apply(logCounts_dgList, 1, min)

logCounts_dgList <-
  logCounts_dgList[(maxTPM >= 2) & ((maxTPM - minTPM) >= 2),]

# Keep the top 500 genes based on variance
keepVariance <-
  order(apply(logCounts_dgList, 1, var), decreasing = T)[1:500]
topVariance_counts <- logCounts_dgList[keepVariance,]

# apply PCA on the transposed data
outputPCA <- prcomp(t(topVariance_counts))

# Calculate the variance of each PC
variancePC <- outputPCA$sdev ^ 2 / sum(outputPCA$sdev ^ 2)
variancePC <- round(variancePC, digits = 2)

variancePC_df <- data.frame(PC = paste0("PC", 1:10),
                            variancePC = variancePC[1:10])

# Create the scree plot
ggplot(variancePC_df, aes(x = PC, y = variancePC)) +
  geom_col() + geom_bar(stat = "identity", fill = "royalblue4") +
  labs(title = "Scree Plot of Principal Components") +
  ylab("Explained variance") + ylim(0, 1)

# Extract the PCs 1 and 2
outputPCA <- as.data.frame(outputPCA$x[, 1:2])

# Add DGEList varaibles to the PCA output
outputPCA[, 3:7] <- sampleInfo[, 1:5]



# Function to assign group based on pattern matching
assign_group <- function(sample, patterns) {
  for (pattern in patterns) {
    if (grepl(pattern, sample)) {
      return(pattern)
    }
  }
  return("Other")  # Default group for samples that don't match any pattern
}

# List of patterns
patterns <- c(
  "AC_F_RT", "AC_M_RT", "AC_R_RT", 
  "MG_F_RT", "MG_M_RT", "MG_R_RT",
  "AC_F_TR", "AC_M_TR", "AC_R_TR", 
  "MG_F_TR", "MG_M_TR", "MG_R_TR"
)


# Apply the function to each sample to create the group column
sampleInfo$Sample <- sapply(sampleInfo$sample, assign_group, patterns)

# Map this new grouping to the PCA output
outputPCA$Sample <- sampleInfo$Sample[match(row.names(outputPCA), sampleInfo$sample)]

# Define colors for each group
group_colors <- c(
  "AC_F_RT" = "green",
  "AC_M_RT" = "yellow",
  "AC_R_RT" = "red",
  "MG_F_RT" = "blue",
  "MG_M_RT" = "pink",
  "MG_R_RT" = "black",
  "AC_F_TR" = "limegreen",
  "AC_M_TR" = "orange",
  "AC_R_TR" = "darkred",
  "MG_F_TR" = "darkblue",
  "MG_M_TR" = "purple",
  "MG_R_TR" = "gray"
)

# Update the ggplot call for PCA plot
ggplot(
  outputPCA,
  aes(
    x = PC1,
    y = PC2,
    color = Sample,  # Use the new grouping variable for color
    shape = replication  # Optionally, you can still shape by original replication
  )
) +
  ggtitle("PCA Plot of Replication by Sample Groups") +
  geom_point(size = 3, alpha = 1) +
  scale_color_manual(values = group_colors, 
                     limits = patterns) +  # Set the order of colors in the legend
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste("PC1: ", variancePC[1], sep = "")) +
  ylab(paste("PC2: ", variancePC[2], sep = ""))

# Update the ggplot call for PCA plot
ggplot(
  outputPCA,
  aes(
    x = PC1,
    y = PC2,
    color = Sample,  # Use the new grouping variable for color
    shape = Condition  # Optionally, you can still shape by original replication
  )
) +
  ggtitle("PCA Plot of Condition by Sample Groups") +
  geom_point(size = 3, alpha = 1) +
  scale_color_manual(values = group_colors, 
                     limits = patterns) +  # Set the order of colors in the legend
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste("PC1: ", variancePC[1], sep = "")) +
  ylab(paste("PC2: ", variancePC[2], sep = ""))

# Update the ggplot call for PCA plot
ggplot(
  outputPCA,
  aes(
    x = PC1,
    y = PC2,
    color = Sample,  # Use the new grouping variable for color
    shape = Region  # Optionally, you can still shape by original replication
  )
) +
  ggtitle("PCA Plot of Regions by Sample Groups") +
  geom_point(size = 3, alpha = 1) +
  scale_color_manual(values = group_colors, 
                     limits = patterns) +  # Set the order of colors in the legend
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste("PC1: ", variancePC[1], sep = "")) +
  ylab(paste("PC2: ", variancePC[2], sep = ""))

# Update the ggplot call for PCA plot
ggplot(
  outputPCA,
  aes(
    x = PC1,
    y = PC2,
    color = Sample,  # Use the new grouping variable for color
    shape = Cell_type  # Optionally, you can still shape by original replication
  )
) +
  ggtitle("PCA Plot of Celltype by Sample Groups") +
  geom_point(size = 3, alpha = 1) +
  scale_color_manual(values = group_colors, 
                     limits = patterns) +  # Set the order of colors in the legend
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste("PC1: ", variancePC[1], sep = "")) +
  ylab(paste("PC2: ", variancePC[2], sep = ""))




# Create a design matrix with our variable of interest
designMatrix <- model.matrix( ~ as.factor(Region), data = sampleInfo)

# Apply the Combat correction of the replicates
outputComBat <- ComBat(
  dat = topVariance_counts,
  batch = as.factor(sampleInfo$replication),
  mod = designMatrix,
  par.prior = F
)

# Set negative values to 0
outputComBat[outputComBat < 0] <- 0

# Apply again PCA
correctedPCA <- prcomp(t(outputComBat))


corrected_varExplained <-
  correctedPCA$sdev ^ 2 / sum(correctedPCA$sdev ^ 2)
corrected_varExplained <- round(corrected_varExplained, digits = 2)

correctedPCA <- as.data.frame(correctedPCA$x[, 1:2])

correctedPCA[, 3:7] <- sampleInfo[, 1:5]

# Map this new grouping to the PCA output
correctedPCA$Sample <- sampleInfo$Sample[match(row.names(correctedPCA), sampleInfo$sample)]

# Update the ggplot call for PCA plot
ggplot(
  correctedPCA,
  aes(
    x = PC1,
    y = PC2,
    color = Sample,  # Use the new grouping variable for color
    shape = replication  # Optionally, you can still shape by original replication
  )
) +
  ggtitle("Combat Corrected PCA Plot of Replication by Sample Groups") +
  geom_point(size = 3, alpha = 1) +
  scale_color_manual(values = group_colors, 
                     limits = patterns) +  # Set the order of colors in the legend
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste("PC1: ", variancePC[1], sep = "")) +
  ylab(paste("PC2: ", variancePC[2], sep = ""))

# Update the ggplot call for PCA plot
ggplot(
  correctedPCA,
  aes(
    x = PC1,
    y = PC2,
    color = Sample,  # Use the new grouping variable for color
    shape = Region  # Optionally, you can still shape by original replication
  )
) +
  ggtitle("Combat Corrected PCA Plot of Region by Sample Groups") +
  geom_point(size = 3, alpha = 1) +
  scale_color_manual(values = group_colors, 
                     limits = patterns) +  # Set the order of colors in the legend
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste("PC1: ", variancePC[1], sep = "")) +
  ylab(paste("PC2: ", variancePC[2], sep = ""))


# Update the ggplot call for PCA plot
ggplot(
  correctedPCA,
  aes(
    x = PC1,
    y = PC2,
    color = Sample,  # Use the new grouping variable for color
    shape = Condition  # Optionally, you can still shape by original replication
  )
) +
  ggtitle("Combat Corrected PCA Plot of Condition by Sample Groups") +
  geom_point(size = 3, alpha = 1) +
  scale_color_manual(values = group_colors, 
                     limits = patterns) +  # Set the order of colors in the legend
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste("PC1: ", variancePC[1], sep = "")) +
  ylab(paste("PC2: ", variancePC[2], sep = ""))

# Update the ggplot call for PCA plot
ggplot(
  correctedPCA,
  aes(
    x = PC1,
    y = PC2,
    color = Sample,  # Use the new grouping variable for color
    shape = Cell_type  # Optionally, you can still shape by original replication
  )
) +
  ggtitle("Combat Corrected PCA Plot of Cell_type by Sample Groups") +
  geom_point(size = 3, alpha = 1) +
  scale_color_manual(values = group_colors, 
                     limits = patterns) +  # Set the order of colors in the legend
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste("PC1: ", variancePC[1], sep = "")) +
  ylab(paste("PC2: ", variancePC[2], sep = ""))




```

```{r}



#
sign_log2FC_AC_FM <- sign_PvalueAC_FM[, 2:3]
rownames(sign_log2FC_AC_FM) <- sign_log2FC_AC_FM$GeneNames
sign_log2FC_AC_FM <- sign_log2FC_AC_FM[, -1]
#
sign_log2FC_AC_FR <- sign_PvalueAC_FR[, 2:3]
rownames(sign_log2FC_AC_FR) <- sign_log2FC_AC_FR$GeneNames
sign_log2FC_AC_FR <- sign_log2FC_AC_FR[, -1]
#
sign_log2FC_AC_MR <- sign_PvalueAC_MR[, 2:3]
rownames(sign_log2FC_AC_MR) <- sign_log2FC_AC_MR$GeneNames
sign_log2FC_AC_MR <- sign_log2FC_AC_MR[, -1]

# Ensure that 'sign_log2FC_AC_FM' is a data frame
sign_log2FC_AC_FM <- as.data.frame(sign_log2FC_AC_FM)
sign_log2FC_AC_FR <- as.data.frame(sign_log2FC_AC_FR)
sign_log2FC_AC_MR <- as.data.frame(sign_log2FC_AC_MR)



#
sign_log2FC_MG_FM <- sign_PvalueMG_FM[, 2:3]
rownames(sign_log2FC_MG_FM) <- sign_log2FC_MG_FM$GeneNames
sign_log2FC_MG_FM <- sign_log2FC_MG_FM[, -1]
#
sign_log2FC_MG_FR <- sign_PvalueMG_FR[, 2:3]
rownames(sign_log2FC_MG_FR) <- sign_log2FC_MG_FR$GeneNames
sign_log2FC_MG_FR <- sign_log2FC_MG_FR[, -1]
#
sign_log2FC_MG_MR <- sign_PvalueMG_MR[, 2:3]
rownames(sign_log2FC_MG_MR) <- sign_log2FC_MG_MR$GeneNames
sign_log2FC_MG_MR <- sign_log2FC_MG_MR[, -1]

# Ensure that 'sign_log2FC_MG_FM' is a data frame
sign_log2FC_MG_FM <- as.data.frame(sign_log2FC_MG_FM)
sign_log2FC_MG_FR <- as.data.frame(sign_log2FC_MG_FR)
sign_log2FC_MG_MR <- as.data.frame(sign_log2FC_MG_MR)


# Now, 'sign_log2FC_MG_FM', 'sign_log2FC_MG_FR', 'sign_log2FC_MG_MR'
# should be a character vector of row names
sign_log2FC_MG_FM <- rownames(sign_log2FC_MG_FM)
sign_log2FC_MG_FR <- rownames(sign_log2FC_MG_FR)
sign_log2FC_MG_MR <- rownames(sign_log2FC_MG_MR)

head(sign_log2FC_MG_FM)
head(sign_log2FC_MG_FR)
head(sign_log2FC_MG_MR)



################### GENE SETS #############

# Create gene sets
gene_set_AC_FM <- as.character(sign_PvalueAC_FM$GeneNames)
gene_set_AC_FR <- as.character(sign_PvalueAC_FR$GeneNames)
gene_set_AC_MR <- as.character(sign_PvalueAC_MR$GeneNames)

gene_set_MG_FM <- as.character(sign_PvalueMG_FM$GeneNames)
gene_set_MG_FR <- as.character(sign_PvalueMG_FR$GeneNames)
gene_set_MG_MR <- as.character(sign_PvalueMG_MR$GeneNames)


# AC_FMR
gene_sets_AC_FMR <-
  list(
    Front_vs_Middle = gene_set_AC_FM,
    Front_vs_Rear = gene_set_AC_FR,
    Middle_vs_Rear = gene_set_AC_MR
  )
plot(
  euler(gene_sets_AC_FMR),
  quantities = TRUE,
  fills = c("red", "green", "yellow")
)

grid.text(
  "Venn Diagram of DGE's for Astrocytes across Front, Middle, and Rear Regions",
  x = 0.5,
  y = 0.95
)
# Find common genes
common_genes_AC <-
  Reduce(intersect,
         list(gene_set_AC_FM, gene_set_AC_FR, gene_set_AC_MR))

# Print common genes
print(common_genes_AC)

# Save common_genes to a file
write.csv(common_genes_AC, "common_genes_AC.csv")

# MG_FMR
gene_sets_MG_FMR <-
  list(
    Front_vs_Middle = gene_set_MG_FM,
    Front_vs_Rear = gene_set_MG_FR,
    Middle_vs_Rear = gene_set_MG_MR
  )
plot(
  euler(gene_sets_MG_FMR),
  quantities = TRUE,
  fills = c("red", "green", "yellow")
)
grid.text(
  "Venn Diagram of DGE's for Microglia across Front, Middle, and Rear Regions",
  x = 0.5,
  y = 0.95
)
# Find common genes
common_genes_MG <-
  Reduce(intersect,
         list(gene_set_MG_FM, gene_set_MG_FR, gene_set_MG_MR))

# Print common genes
print(common_genes_MG)

# Save common_genes to a file
write.csv(common_genes_MG, "common_genes_MG.csv")

## FM
# Create a named list of gene sets
gene_sets_AC_MG_FM <-
  list(Astrocyte_Front_vs_Middle = gene_set_AC_FM,
       Microglia_Front_vs_Middle = gene_set_MG_FM)

# Create an Euler diagram (a type of Venn diagram) with custom plot parameters
plot(
  euler(gene_sets_AC_MG_FM),
  quantities = TRUE,
  fills = c("red", "green")
)
grid.text(
  "Venn Diagram of DGE's for Astrocytes and Microglia Front vs Middle",
  x = 0.5,
  y = 0.90
)
# Find common genes
common_genes_MG_AC_FM <-
  Reduce(intersect, list(gene_set_MG_FM, gene_set_AC_FM))

# Print common genes
print(common_genes_MG_AC_FM)


## FR
# Create a named list of gene sets
gene_sets_AC_MG_FR <-
  list(Astrocyte_Front_vs_Rear = gene_set_AC_FR,
       Microglia_Front_vs_Rear = gene_set_MG_FR)

# Create an Euler diagram (a type of Venn diagram) with custom plot parameters
plot(
  euler(gene_sets_AC_MG_FR),
  quantities = TRUE,
  fills = c("red", "green")
)
grid.text(
  "Venn Diagram of DGE's for Astrocytes and Microglia Front vs Rear",
  x = 0.5,
  y = 0.93
)

# Find common genes
common_genes_MG_AC_FR <-
  Reduce(intersect, list(gene_set_MG_FR, gene_set_AC_FR))

# Print common genes
print(common_genes_MG_AC_FR)

## MR
# Create a named list of gene sets
gene_sets_AC_MG_MR <-
  list(Astrocyte_Middle_vs_Rear = gene_set_AC_MR,
       Microglia_Middle_vs_Rear = gene_set_MG_MR)

# Create an Euler diagram (a type of Venn diagram) with custom plot parameters
plot(
  euler(gene_sets_AC_MG_MR),
  quantities = TRUE,
  fills = c("red", "green")
)
grid.text(
  "Venn Diagram of DGE's for Astrocytes and Microglia Middle vs Rear",
  x = 0.5,
  y = 0.95
)
# Find common genes
common_genes_MG_AC_MR <-
  Reduce(intersect, list(gene_set_MG_MR, gene_set_AC_MR))

# Print common genes
print(common_genes_MG_AC_MR)

# Save all the intersected genes
write.csv(common_genes_MG_AC_FM, "common_genes_MG_AC_FM.csv")
write.csv(common_genes_MG_AC_FR, "common_genes_MG_AC_FR.csv")
write.csv(common_genes_MG_AC_MR, "common_genes_MG_AC_MR.csv")

```


```{r}
# read the data file with all samples
rawCounts <-
  read.table("raw_counts_cpm_after_norm_reorganised_220722.txt", header = TRUE)


# # USing biomaRt annotation  method.
# ensembl <- useMart("ensembl")
# 
# listDatasets(ensembl)
# 
# ensembl_mouse <- useDataset("mmusculus_gene_ensembl", mart = ensembl)
# 
# mouse_gene_ids <- rawCounts$ensembl_id
# mouse_gene_info <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "chromosome_name", "start_position", "end_position", "gene_biotype", "strand"),
#                          filters = "ensembl_gene_id",
#                          values = mouse_gene_ids,
#                          mart = ensembl_mouse)

# Changing name of column too merge the values with a matching column.
colnames(mouse_gene_info)[c(1)] <- c("ensembl_id")
# Merging the two dataframes into one dataframe leaving out one
rawCounts <- merge(rawCounts, mouse_gene_info, c(1))

rawCounts <- rawCounts[,-c(1, 38:42)]

colnames(rawCounts)[c(36)] <-  c("GeneNames")

# keep only the first occurrence of each duplicated gene name
rawCounts <- rawCounts[!duplicated(rawCounts$GeneNames),]

# add gene name to  row names 
row.names(rawCounts) <-  rawCounts$GeneNames
rawCounts <- rawCounts[,-36]

# Keep only rows where all values are above 1
rawCounts <- rawCounts[apply(rawCounts, 1, function(x) all(x > 1)), ]


# separating the columns if ribotag and totalRNA  
# into  separated dataframe object for the rawCounts and sampleInfo
rawCounts_RT <- rawCounts[, grep("_RT", colnames(rawCounts))]
#rawCounts_TR <- rawCounts[, grep("_TR", colnames(rawCounts))]

rawCounts_RT <- rawCounts_RT[order(colnames(rawCounts_RT))]

rawCounts_RT$AC_F <- rowMeans(rawCounts_RT[,1:2])
rawCounts_RT$AC_M <- rowMeans(rawCounts_RT[,3:5])
rawCounts_RT$AC_R <- rowMeans(rawCounts_RT[,6:8])

rawCounts_RT$MG_F <- rowMeans(rawCounts_RT[,9:11])
rawCounts_RT$MG_M <- rowMeans(rawCounts_RT[,12:14])
rawCounts_RT$MG_R <- rowMeans(rawCounts_RT[,15:17])

rawCounts_RT <- rawCounts_RT[,-c(1:17)]


#rawCounts_TR <- rawCounts_TR[order(colnames(rawCounts_TR))]


# making function for calculating z score since scale function is not ideal
calc_zscore <- function(df) {
  result <- (df - mean(df)) / sd(df)
  return(result)
}

# calculating the Z-score of ribotag over the Z-score of totalRNA
# calculating the z-score
Z_score_RT <- apply(rawCounts_RT, 1, calc_zscore)
Z_score_RT <- data.frame(t(Z_score_RT))

Z_score_RT <- Z_score_RT[apply(Z_score_RT, 1, function(x) {all(x > -1.1 & x < 1.1)}),]

sample_ratios <- data.frame(
  AC_F = Z_score_RT[, grep("AC_F", colnames(Z_score_RT))],
  AC_M = Z_score_RT[, grep("AC_M", colnames(Z_score_RT))],
  AC_R = Z_score_RT[, grep("AC_R", colnames(Z_score_RT))],
  MG_F = Z_score_RT[, grep("MG_F", colnames(Z_score_RT))],
  MG_M = Z_score_RT[, grep("MG_M", colnames(Z_score_RT))],
  MG_R = Z_score_RT[, grep("MG_R", colnames(Z_score_RT))]
)

rownames(sample_ratios) <- rownames(Z_score_RT)
sample_ratios <- na.omit(sample_ratios)
# Convert the dataframe to a binary format: 1 if not NA, 0 otherwise
binary_matrix <- as.data.frame(lapply(sample_ratios, function(x) as.integer(x)))

# Preserve the row names in the binary matrix
row.names(binary_matrix) <- row.names(sample_ratios)


# Create the UpSet plot
upset(
  binary_matrix,
  sets = c("AC_F", "AC_M", "AC_R", "MG_F", "MG_M", "MG_R"),
  order.by = "freq",
  mainbar.y.label = "Intersection Size",
  sets.bar.color = "black"
)

Matrix_Samples <- as.matrix(sample_ratios)
Matrix_Samples <- na.omit(Matrix_Samples)
pheatmap(Matrix_Samples)

Upset_heatmap <- pheatmap(
  t(Matrix_Samples),
  cluster_cols = T,
  cluster_rows = F,
  main = "Z-score of each celltype's average across regions",
  fontsize = 5,
  cellwidth = 4.5,
  cellheight = 30
  
)

ggsave(
  filename = "UpsetheatmapPlot1.png",
  plot = Upset_heatmap,
  width = 20,
  height = 13
)

binary_matrix$GeneNames <- rownames(binary_matrix)
write_xlsx(binary_matrix, "binary_matrix_upsetPlot.xlsx")
sample_ratios$GeneNames <- rownames(sample_ratios)
write_xlsx(sample_ratios,
"UpsetPlot_GeneList.xlsx")

```


```{r}

# read the data file with all samples
rawCounts <-
  read.table("raw_counts_cpm_after_norm_reorganised_220722.txt", header = TRUE)


# Load sample information data
sampleInfo <- read.table("Sample-info_220722.txt", header = TRUE)

# Load cell type marker genes data from an Excel file
markerGenes <- read_excel("Celltype_markergenes.xlsx")


# ensembl <- useMart("ensembl")
# listDatasets(ensembl)
# 
# ensembl_mouse <- useDataset("mmusculus_gene_ensembl", mart = ensembl)
# 
# mouse_gene_ids <- rawCounts$ensembl_id
# mouse_gene_info <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "chromosome_name", "start_position", "end_position", "gene_biotype", "strand"),
#                          filters = "ensembl_gene_id",
#                          values = mouse_gene_ids,
#                          mart = ensembl_mouse)


# Changing name of column too merge the values with a matching column.
colnames(mouse_gene_info)[c(1)] <- c("ensembl_id")
# Merging the two dataframes into one dataframe leaving out one
rawCounts <- merge(rawCounts, mouse_gene_info)

rawCounts <- rawCounts[,-c(1, 38:42)]

colnames(rawCounts)[c(36)] <-  c("GeneNames")
########
# keep only the first occurrence of each duplicated gene name
rawCounts <- rawCounts[!duplicated(rawCounts$GeneNames),]


row.names(rawCounts) <-  rawCounts$GeneNames
rawCounts <- rawCounts[,-36]

rawCounts_RT <- rawCounts[,1:17]
# Plot two genes only RT

# making function for calculating z score since scale function is not ideal
calc_zscore <- function(df) {
  result <- (df - mean(df)) / sd(df)
  return(result)
}

Z_score_RT <- apply(rawCounts_RT, 1, calc_zscore)
Z_score_RT <- as.data.frame(t(Z_score_RT))
TwoGenes <- data.frame(TwoGenes <- Z_score_RT[c("Ndufb4c", "Ndufb11"), ])
TwoGenes$genenames <- rownames(TwoGenes)

# Get my data into a long format
TwoGenes_boxPlot <-
  gather(TwoGenes, key = "samples", value = "values",-c(genenames))

# Filter samples by the specified regions
regions <- c("_F", "_M", "_R")
TwoGenes_boxPlot <- TwoGenes_boxPlot %>%
  filter(str_detect(samples, paste0("AC", "|", "MG")) &
           str_detect(samples, paste(regions, collapse = "|")))

# Assign groups based on the cell type
TwoGenes_boxPlot$group <-
  ifelse(str_detect(TwoGenes_boxPlot$samples, "AC"),
         "Astrocytes",
         "Microglia")

# Function to create a boxplot for the specified regions
create_boxplot <- function(region1, region2) {
  # Filter samples by the specified regions
  TwoGenes_boxPlot_filtered <- TwoGenes_boxPlot %>%
    filter(str_detect(samples, paste(c(region1, region2), collapse = "|")))
  
  # Create the boxplot
  ggplot(TwoGenes_boxPlot_filtered,
         aes(x = genenames, y = values, fill = group)) +
    geom_boxplot() +
    scale_fill_manual(values = c("green", "red")) +
    theme_light() +
    ggtitle(paste("Boxplot for", region1, "and", region2, "regions"))
}

# Create separate boxplots for the specified region pairs
boxplot_front_middle <- create_boxplot("_F", "_M")
boxplot_front_rare <- create_boxplot("_F", "_R")
boxplot_middle_rare <- create_boxplot("_M", "_R")

# Display the boxplots
boxplot_front_middle
boxplot_front_rare
boxplot_middle_rare




# Function to assign region labels to the samples
assign_region <- function(samples) {
  if (str_detect(samples, "_F")) {
    return("Front")
  } else if (str_detect(samples, "_M")) {
    return("Middle")
  } else if (str_detect(samples, "_R")) {
    return("Rare")
  } else {
    return(NA)
  }
}

# Assign region labels to the samples
TwoGenes_boxPlot$region <-
  sapply(TwoGenes_boxPlot$samples, assign_region)

# Filter samples by the specified regions
TwoGenes_boxPlot_filtered <- TwoGenes_boxPlot %>%
  filter(str_detect(samples, paste(regions, collapse = "|")))

# Create the boxplot with separate facets for each region pair
ggplot(TwoGenes_boxPlot_filtered,
       aes(x = genenames, y = values, fill = group)) +
  geom_boxplot() +
  scale_fill_manual(values = c("green", "red")) +
  theme_light() +
  facet_wrap( ~ region, ncol = 3) +
  ggtitle("Boxplots for gene expression per region for astrocytes and microglia, Ribotag data")


```
