---
title: "DGE analysis of Astrocyte and Microglia across different cerebral regions"
author: "Link√∂ping University, Joseph Filip Agi Maqdissi" 
date: "`r Sys.Date()`"
output: html_document
---

This is a protocol for the analysis of the data preformed in R and the packages being used in Rstudio. Read through the code and comments carefully. The protocole displayes the output in the terminal and the plots generated by the code. 

If there is any questions feel free to mail: Josag897@student.liu.se

```{r}

# loading necessary library
library(pathview)
library(clusterProfiler)
library(dplyr)
library(limma)
library(edgeR)
library(org.Mm.eg.db)
library(AnnotationDbi)
library(tidyverse)
library(biomaRt)
library(readxl)
library(affy)
library(ggplot2)
library(sva)
library(gplots)
library(VennDiagram)
library(eulerr)
library(UpSetR)

#  Load file with Counts per million after normalization data
rawCounts <-
  read.table("raw_counts_cpm_after_norm_reorganised_220722.txt", header = T)

# Converting ensemblID to symbol
rawCounts$GeneNames <- mapIds(
  org.Mm.eg.db,
  keys = rawCounts$ensembl_id,
  keytype = "ENSEMBL",
  column = "SYMBOL"
)


# Removing NA values
rawCounts <- na.omit(rawCounts)


# Removing ENsemblID column
rawCounts <- rawCounts[, -1]
# Order the columns
rawCounts <- rawCounts[order(colnames(rawCounts))]


# Build a separated dataframe for the ribotag and totalRNA for astrocyte and microglia sep.
AC_RT <-
  subset(rawCounts,
         !(rownames(rawCounts) %in% "16712"),
         select = c(1, 2, 6, 7, 8, 12, 13, 14, 18))
AC_TR <-
  subset(rawCounts,
         !(rownames(rawCounts) %in% "16712"),
         select = c(3, 4, 5, 9, 10, 11, 15, 16, 17, 18))

MG_RT <-
  subset(rawCounts,
         !(rownames(rawCounts) %in% "16712"),
         select = c(19, 20, 21, 25, 26, 27, 31, 32, 33, 18))
MG_TR <-
  subset(rawCounts,
         !(rownames(rawCounts) %in% "16712"),
         select = c(22, 23, 24, 28, 29, 30, 34, 35, 36, 18))


# Check the variance of the gene expression data
AC_RT_var <- sapply(AC_RT, var)
AC_TR_var <- sapply(AC_TR, var)

MG_RT_var <- sapply(MG_RT, var)
MG_TR_var <- sapply(MG_TR, var)

# Plot the variance distribution to check if it is homogeneous
hist(AC_RT_var,
     main = "Histogram of Ribotag Astrocyte Values Distribution",
     xlab = "Values",
     col = "lightblue") # plot histogram
hist(AC_TR_var,
     main = "Histogram of totalRNA Astrocyte Values Distribution",
     xlab = "Values",
     col = "lightblue") # plot histogram

hist(MG_RT_var,
     main = "Histogram of Ribotag Microglia Values Distribution",
     xlab = "Values",
     col = "lightblue") # plot histogram
hist(MG_TR_var,
     main = "Histogram of totalRNA Microglia Values Distribution",
     xlab = "Values",
     col = "lightblue") # plot histogram


# Creating new Objects
# AC
AC_F_rep1 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(1, 3, 18))
AC_F_rep2 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(2, 4, 18))

AC_M_rep1 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(6, 9, 18))
AC_M_rep2 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(7, 10, 18))
AC_M_rep3 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(8, 11, 18))

AC_R_rep1 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(12, 15, 18))
AC_R_rep2 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(13, 16, 18))
AC_R_rep3 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(14, 17, 18))

# MG
MG_F_rep1 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(19, 22, 18))
MG_F_rep2 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(20, 23, 18))
MG_F_rep3 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(21, 24, 18))

MG_M_rep1 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(25, 28, 18))
MG_M_rep2 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(26, 29, 18))
MG_M_rep3 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(27, 30, 18))

MG_R_rep1 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(31, 34, 18))
MG_R_rep2 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(32, 35, 18))
MG_R_rep3 <-
  subset(rawCounts,!(rownames(rawCounts) %in% "16712"), select = c(33, 36, 18))



# divide the RT over TR  and then preform a KEGG enrichment

# AC
# Front
AC_F_rep1$RT_TR_F_rep1 <-
  AC_F_rep1$AC_F_RT_rep1 / AC_F_rep1$AC_F_TR_rep1

AC_F_rep2$RT_TR_F_rep2 <-
  AC_F_rep2$AC_F_RT_rep2 / AC_F_rep2$AC_F_TR_rep2

# Middle
AC_M_rep1$RT_TR_M_rep1 <-
  AC_M_rep1$AC_M_RT_rep1 / AC_M_rep1$AC_M_TR_rep1

AC_M_rep2$RT_TR_M_rep2 <-
  AC_M_rep2$AC_M_RT_rep2 / AC_M_rep2$AC_M_TR_rep2

AC_M_rep3$RT_TR_M_rep3 <-
  AC_M_rep3$AC_M_RT_rep3 / AC_M_rep3$AC_M_TR_rep3

# Rare
AC_R_rep1$RT_TR_R_rep1 <-
  AC_R_rep1$AC_R_RT_rep1 / AC_R_rep1$AC_R_TR_rep1

AC_R_rep2$RT_TR_R_rep2 <-
  AC_R_rep2$AC_R_RT_rep2 / AC_R_rep2$AC_R_TR_rep2

AC_R_rep3$RT_TR_R_rep3 <-
  AC_R_rep3$AC_R_RT_rep3 / AC_R_rep3$AC_R_TR_rep3

# MG
# Front
MG_F_rep1$RT_TR_F_rep1 <-
  MG_F_rep1$MG_F_RT_rep1 / MG_F_rep1$MG_F_TR_rep1

MG_F_rep2$RT_TR_F_rep2 <-
  MG_F_rep2$MG_F_RT_rep2 / MG_F_rep2$MG_F_TR_rep2

# Middle
MG_M_rep1$RT_TR_M_rep1 <-
  MG_M_rep1$MG_M_RT_rep1 / MG_M_rep1$MG_M_TR_rep1

MG_M_rep2$RT_TR_M_rep2 <-
  MG_M_rep2$MG_M_RT_rep2 / MG_M_rep2$MG_M_TR_rep2

MG_M_rep3$RT_TR_M_rep3 <-
  MG_M_rep3$MG_M_RT_rep3 / MG_M_rep3$MG_M_TR_rep3

# Rare
MG_R_rep1$RT_TR_R_rep1 <-
  MG_R_rep1$MG_R_RT_rep1 / MG_R_rep1$MG_R_TR_rep1

MG_R_rep2$RT_TR_R_rep2 <-
  MG_R_rep2$MG_R_RT_rep2 / MG_R_rep2$MG_R_TR_rep2

MG_R_rep3$RT_TR_R_rep3 <-
  MG_R_rep3$MG_R_RT_rep3 / MG_R_rep3$MG_R_TR_rep3
# Preforming a merge of the different ribotag/totalRNA values preforming enrichment on them
# to find the most expressed genes in their enrichment ratio scores

# AC
# Front
RatioData_Actrocytes <-
  subset(AC_F_rep1,!(rownames(AC_F_rep1) %in% "16712"), select = RT_TR_F_rep1)

RatioData_Actrocytes$RT_TR_F_rep2 <- AC_F_rep2$RT_TR_F_rep2

# Middle
RatioData_Actrocytes$RT_TR_M_rep1 <- AC_M_rep1$RT_TR_M_rep1
RatioData_Actrocytes$RT_TR_M_rep2 <- AC_M_rep2$RT_TR_M_rep2
RatioData_Actrocytes$RT_TR_M_rep3 <- AC_M_rep3$RT_TR_M_rep3

# Rare
RatioData_Actrocytes$RT_TR_R_rep1 <- AC_R_rep1$RT_TR_R_rep1
RatioData_Actrocytes$RT_TR_R_rep2 <- AC_R_rep2$RT_TR_R_rep2
RatioData_Actrocytes$RT_TR_R_rep3 <- AC_R_rep3$RT_TR_R_rep3

RatioData_Actrocytes$GeneNames <- AC_F_rep1$GeneNames

RatioData_Actrocytes <- na.omit(RatioData_Actrocytes)

# Remove inf values
RatioData_Actrocytes[RatioData_Actrocytes == Inf] <- 0
# RatioData_Actrocytes <- RatioData_Actrocytes %>% filter_all(all_vars(!is.infinite(.)))

# MG
# Front
RatioData_Microglia <-
  subset(MG_F_rep1,!(rownames(MG_F_rep1) %in% "16712"), select = RT_TR_F_rep1)

RatioData_Microglia$RT_TR_F_rep2 <- MG_F_rep2$RT_TR_F_rep2

# Middle
RatioData_Microglia$RT_TR_M_rep1 <- MG_M_rep1$RT_TR_M_rep1
RatioData_Microglia$RT_TR_M_rep2 <- MG_M_rep2$RT_TR_M_rep2
RatioData_Microglia$RT_TR_M_rep3 <- MG_M_rep3$RT_TR_M_rep3

# Rare
RatioData_Microglia$RT_TR_R_rep1 <- MG_R_rep1$RT_TR_R_rep1
RatioData_Microglia$RT_TR_R_rep2 <- MG_R_rep2$RT_TR_R_rep2
RatioData_Microglia$RT_TR_R_rep3 <- MG_R_rep3$RT_TR_R_rep3

RatioData_Microglia$GeneNames <- MG_F_rep1$GeneNames

RatioData_Microglia <- na.omit(RatioData_Microglia)

# Remove inf values
RatioData_Microglia[RatioData_Microglia == Inf] <- 0
# RatioData_Microglia <- RatioData_Microglia %>% filter_all(all_vars(!is.infinite(.)))

# Make object to  store p-values
PvalueAC_FM <- data.frame(Pvalue = rep(0, nrow(RatioData_Actrocytes)))
PvalueAC_FR <- data.frame(Pvalue = rep(0, nrow(RatioData_Actrocytes)))
PvalueAC_MR <- data.frame(Pvalue = rep(0, nrow(RatioData_Actrocytes)))

PvalueMG_FM <- data.frame(Pvalue = rep(0, nrow(RatioData_Microglia)))
PvalueMG_FR <- data.frame(Pvalue = rep(0, nrow(RatioData_Microglia)))
PvalueMG_MR <- data.frame(Pvalue = rep(0, nrow(RatioData_Microglia)))


for (i in 1:nrow(RatioData_Actrocytes)) {
  PvalueAC_FM$Pvalue[i] <-
    t.test(RatioData_Actrocytes[i, 1:2], RatioData_Actrocytes[i, 3:5], alternative = "two.sided")$p.value
  PvalueAC_FR$Pvalue[i] <-
    t.test(RatioData_Actrocytes[i, 1:2], RatioData_Actrocytes[i, 6:8], alternative = "two.sided")$p.value
  PvalueAC_MR$Pvalue[i] <-
    t.test(RatioData_Actrocytes[i, 3:5], RatioData_Actrocytes[i, 6:8], alternative = "two.sided")$p.value
}

for (i in 1:nrow(RatioData_Microglia)) {
  PvalueMG_FM$Pvalue[i] <-
    t.test(RatioData_Microglia[i, 1:2], RatioData_Microglia[i, 3:5], alternative = "two.sided")$p.value
  PvalueMG_FR$Pvalue[i] <-
    t.test(RatioData_Microglia[i, 1:2], RatioData_Microglia[i, 6:8], alternative = "two.sided")$p.value
  PvalueMG_MR$Pvalue[i] <-
    t.test(RatioData_Microglia[i, 3:5], RatioData_Microglia[i, 6:8], alternative = "two.sided")$p.value
}

# add the column genename
PvalueAC_FM$GeneNames <- RatioData_Actrocytes$GeneNames
PvalueAC_FR$GeneNames <- RatioData_Actrocytes$GeneNames
PvalueAC_MR$GeneNames <- RatioData_Actrocytes$GeneNames

PvalueMG_FM$GeneNames <- RatioData_Microglia$GeneNames
PvalueMG_FR$GeneNames <- RatioData_Microglia$GeneNames
PvalueMG_MR$GeneNames <- RatioData_Microglia$GeneNames


# make a FClog column into same pValobject and filter based onm p-value
PvalueAC_FM$log2FC <-
  rowMeans(RatioData_Actrocytes[, 1:2]) - rowMeans(RatioData_Actrocytes[, 3:5])
PvalueAC_FR$log2FC <-
  rowMeans(RatioData_Actrocytes[, 1:2]) - rowMeans(RatioData_Actrocytes[, 6:8])
PvalueAC_MR$log2FC <-
  rowMeans(RatioData_Actrocytes[, 3:5]) - rowMeans(RatioData_Actrocytes[, 6:8])

PvalueMG_FM$log2FC <-
  rowMeans(RatioData_Microglia[, 1:2]) - rowMeans(RatioData_Microglia[, 3:5])
PvalueMG_FR$log2FC <-
  rowMeans(RatioData_Microglia[, 1:2]) - rowMeans(RatioData_Microglia[, 6:8])
PvalueMG_MR$log2FC <-
  rowMeans(RatioData_Microglia[, 3:5]) - rowMeans(RatioData_Microglia[, 6:8])


#Removing the P-values over 0.05 for all p-values objects
#AC
sign_PvalueAC_FM <-
  PvalueAC_FM[PvalueAC_FM$Pvalue < 0.05, , drop = F]
sign_PvalueAC_FM <- na.omit(sign_PvalueAC_FM)

sign_PvalueAC_FR <-
  PvalueAC_FR[PvalueAC_FR$Pvalue < 0.05, , drop = F]
sign_PvalueAC_FR <- na.omit(sign_PvalueAC_FR)

sign_PvalueAC_MR <-
  PvalueAC_MR[PvalueAC_MR$Pvalue < 0.05, , drop = F]
sign_PvalueAC_MR <- na.omit(sign_PvalueAC_MR)

# MG
sign_PvalueMG_FM <-
  PvalueMG_FM[PvalueMG_FM$Pvalue < 0.05, , drop = F]
sign_PvalueMG_FM <- na.omit(sign_PvalueMG_FM)

sign_PvalueMG_FR <-
  PvalueMG_FR[PvalueMG_FR$Pvalue < 0.05, , drop = F]
sign_PvalueMG_FR <- na.omit(sign_PvalueMG_FR)

sign_PvalueMG_MR <-
  PvalueMG_MR[PvalueMG_MR$Pvalue < 0.05, , drop = F]
sign_PvalueMG_MR <- na.omit(sign_PvalueMG_MR)


# Convert gene symbols to ENTREZIDs
Deg_AC_FM <-
  bitr(
    sign_PvalueAC_FM$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )
Deg_AC_FR <-
  bitr(
    sign_PvalueAC_FR$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )
Deg_AC_MR <-
  bitr(
    sign_PvalueAC_MR$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )

Deg_MG_FM <-
  bitr(
    sign_PvalueMG_FM$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )
Deg_MG_FR <-
  bitr(
    sign_PvalueMG_FR$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )
Deg_MG_MR <-
  bitr(
    sign_PvalueMG_MR$GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  )

# One gene was converted twice by bitr in the sign_PvslueMG_FR object, thus it contains duplicates
# Remove one of the duplicates
sign_PvalueMG_FR <-
  sign_PvalueMG_FR[!duplicated(sign_PvalueMG_FR$GeneNames),]


##
# TOO do
# Pathway analysis


# create a background gene file on all of the genes
Backround_Genes <-
  data.frame(rawCounts$GeneNames) # create a data frame with the names of all measured genes

Backround_Genes <-
  bitr(
    Backround_Genes$rawCounts.GeneNames,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = "org.Mm.eg.db"
  ) # convert symbol to ENTREZID

table(duplicated(Backround_Genes$ENTREZID)) #check if duplicated entrez ID


# AC
keggAstrocyte_FM <- enrichKEGG(
  gene = Deg_AC_FM$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggAstrocyte_FM <-
  setReadable(keggAstrocyte_FM, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggAstrocyte_summary_FM <- keggAstrocyte_FM@result
keggAstrocyte_summary_FM <-
  keggAstrocyte_summary_FM[keggAstrocyte_summary_FM$pvalue < 0.05, , drop = F]

#
keggAstrocyte_FR <- enrichKEGG(
  gene = Deg_AC_FR$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggAstrocyte_FM <-
  setReadable(keggAstrocyte_FR, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggAstrocyte_summary_FR <- keggAstrocyte_FR@result
keggAstrocyte_summary_FR <-
  keggAstrocyte_summary_FR[keggAstrocyte_summary_FR$p.adjust < 0.05, , drop = F]

#
keggAstrocyte_MR <- enrichKEGG(
  gene = Deg_AC_MR$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggAstrocyte_MR <-
  setReadable(keggAstrocyte_MR, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggAstrocyte_summary_MR <- keggAstrocyte_MR@result
keggAstrocyte_summary_MR <-
  keggAstrocyte_summary_MR[keggAstrocyte_summary_MR$pvalue < 0.05, , drop = F]

# KEGG pathway analysis
# MG
keggMicroglia_FM <- enrichKEGG(
  gene = Deg_MG_FM$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggMicroglia_FM <-
  setReadable(keggMicroglia_FM, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggMicroglia_summary_FM <- keggMicroglia_FM@result
keggMicroglia_summary_FM <-
  keggMicroglia_summary_FM[keggMicroglia_summary_FM$pvalue < 0.05, , drop = F]

#
keggMicroglia_FR <- enrichKEGG(
  gene = Deg_MG_FR$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggMicroglia_FM <-
  setReadable(keggMicroglia_FR, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggMicroglia_summary_FR <- keggMicroglia_FR@result
keggMicroglia_summary_FR <-
  keggMicroglia_summary_FR[keggMicroglia_summary_FR$pvalue < 0.05, , drop = F]

#
keggMicroglia_MR <- enrichKEGG(
  gene = Deg_MG_MR$ENTREZID,
  organism = "mmu",
  keyType = "kegg",
  universe = Backround_Genes$ENTREZID,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 1
)


keggMicroglia_MR <-
  setReadable(keggMicroglia_MR, OrgDb = org.Mm.eg.db, keyType = "ENTREZID")


keggMicroglia_summary_MR <- keggMicroglia_MR@result
keggMicroglia_summary_MR <-
  keggMicroglia_summary_MR[keggMicroglia_summary_MR$pvalue < 0.05, , drop = F]
########################

create_custom_dot_plot <- function(Kegg_result) {
  # Create a new column 'geneRatioDecimal' in our dataframe which holds the numeric ratio
  Kegg_result$GeneRatioDecimal <-
    as.numeric(unlist(lapply(strsplit(as.character(Kegg_result$GeneRatio), "/"), function(x)
      as.numeric(x[1]) / as.numeric(x[2]))))
  
  dot_plot <-
    ggplot(Kegg_result,
           aes(
             x = reorder(Description,-GeneRatioDecimal),
             y = GeneRatioDecimal,
             color = pvalue,
             size = Count
           )) +
    geom_point() +
    scale_color_gradient(low = "blue", high = "red") +
    theme_minimal() +
    theme(
      axis.title.x = element_blank(),
      axis.text.x = element_text(
        size = 8,
        angle = 75,
        hjust = 1
      ),
      plot.title = element_text(size = 12, hjust = 0.5),
      panel.background = element_rect(fill = "gray90"),
      plot.background = element_rect(fill = "white")
    ) +
    labs(title = "Kegg Enrichment plot",
         y = "GeneRatio",
         x = "Description")
  
  
  return(dot_plot)
}

dot_plot_AC_FM <- create_custom_dot_plot(keggAstrocyte_summary_FM)
dot_plot_AC_FR <- create_custom_dot_plot(keggAstrocyte_summary_FR)
dot_plot_AC_MR <- create_custom_dot_plot(keggAstrocyte_summary_MR)

dot_plot_MG_FM <- create_custom_dot_plot(keggMicroglia_summary_FM)
dot_plot_MG_FR <- create_custom_dot_plot(keggMicroglia_summary_FR)
dot_plot_MG_MR <- create_custom_dot_plot(keggMicroglia_summary_MR)

ggsave(
  filename = "dot_plot_AC_FM.png",
  plot = dot_plot_AC_FM,
  width = 10,
  height = 12,
  dpi = 300
)
ggsave(
  filename = "dot_plot_AC_FR.png",
  plot = dot_plot_AC_FR,
  width = 10,
  height = 12,
  dpi = 300
)
ggsave(
  filename = "dot_plot_AC_MR.png",
  plot = dot_plot_AC_MR,
  width = 10,
  height = 12,
  dpi = 300
)

ggsave(
  filename = "dot_plot_MG_FM.png",
  plot = dot_plot_MG_FM,
  width = 10,
  height = 12,
  dpi = 300
)
ggsave(
  filename = "dot_plot_MG_FR.png",
  plot = dot_plot_MG_FR,
  width = 10,
  height = 12,
  dpi = 300
)
ggsave(
  filename = "dot_plot_MG_MR.png",
  plot = dot_plot_MG_MR,
  width = 10,
  height = 12,
  dpi = 300
)


#AC
log2fc_AC_FM <- data.frame(sign_PvalueAC_FM$log2FC)
rownames(log2fc_AC_FM) <- Deg_AC_FM$ENTREZID

log2fc_AC_FR <- data.frame(sign_PvalueAC_FR$log2FC)
rownames(log2fc_AC_FR) <- Deg_AC_FR$ENTREZID

log2fc_AC_MR <- data.frame(sign_PvalueAC_MR$log2FC)
rownames(log2fc_AC_MR) <- Deg_AC_MR$ENTREZID

# MG
log2fc_MG_FM <- data.frame(sign_PvalueMG_FM$log2FC)
rownames(log2fc_MG_FM) <- Deg_MG_FM$ENTREZID

log2fc_MG_FR <- data.frame(sign_PvalueMG_FR$log2FC)
rownames(log2fc_MG_FR) <- Deg_MG_FR$ENTREZID

log2fc_MG_MR <- data.frame(sign_PvalueMG_MR$log2FC)
rownames(log2fc_MG_MR) <- Deg_MG_MR$ENTREZID

# Pathview analysis
# AC_FM
pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu00980",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu05208",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu05226",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu05221",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu00970",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FM,
  pathway.id = "mmu00910",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

# MG_FM
pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu04141",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu04921",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu04610",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu04152",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu04310",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FM,
  pathway.id = "mmu05231",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

# AC_FR
pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu04260",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu00190",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu04714",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu04666",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu00020",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_FR,
  pathway.id = "mmu05208",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

# MG_FR
pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu04142",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu00190",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu00100",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu03420",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu05167",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_FR,
  pathway.id = "mmu00280",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)



# AC_MR
pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu00640",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu04724",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu00280",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu03010",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu05100",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_AC_MR,
  pathway.id = "mmu00650",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

# MG_MR
pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu05167",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu05203",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu05163",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu04926",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu04612",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)

pathview(
  gene.data = log2fc_MG_MR,
  pathway.id = "mmu05170",
  low = "red",
  mid = "grey",
  high = "green",
  species = "mmu",
  limit = c(-10, 10)
)



# AC
num_log2fc_AC_FM <- sign_PvalueAC_FM$log2FC
names(num_log2fc_AC_FM) <- Deg_AC_FM$ENTREZID

num_log2fc_AC_FR <- sign_PvalueAC_FR$log2FC
names(num_log2fc_AC_FR) <- Deg_AC_FR$ENTREZID

num_log2fc_AC_MR <- sign_PvalueAC_MR$log2FC
names(num_log2fc_AC_MR) <- Deg_AC_MR$ENTREZID


# MG
num_log2fc_MG_FM <- sign_PvalueMG_FM$log2FC
names(num_log2fc_MG_FM) <- Deg_MG_FM$ENTREZID

num_log2fc_MG_FR <- sign_PvalueMG_FR$log2FC
names(num_log2fc_MG_FR) <- Deg_MG_FR$ENTREZID

num_log2fc_MG_MR <- sign_PvalueMG_MR$log2FC
names(num_log2fc_MG_MR) <- Deg_MG_MR$ENTREZID

# AC
Sort_log2fc_AC_FM <- sort(num_log2fc_AC_FM, decreasing = TRUE)
Sort_log2fc_AC_FR <- sort(num_log2fc_AC_FR, decreasing = TRUE)
Sort_log2fc_AC_MR <- sort(num_log2fc_AC_MR, decreasing = TRUE)

# MG
Sort_log2fc_MG_FM <- sort(num_log2fc_MG_FM, decreasing = TRUE)
Sort_log2fc_MG_FR <- sort(num_log2fc_MG_FR, decreasing = TRUE)
Sort_log2fc_MG_MR <- sort(num_log2fc_MG_MR, decreasing = TRUE)



# AC
gseaKEGG_AC_FM <- gseKEGG(
  geneList = Sort_log2fc_AC_FM,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus
gseaKEGG_AC_FR <- gseKEGG(
  geneList = Sort_log2fc_AC_FR,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus
gseaKEGG_AC_MR <- gseKEGG(
  geneList = Sort_log2fc_AC_MR,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus

# MG
gseaKEGG_MG_FM <- gseKEGG(
  geneList = Sort_log2fc_MG_FM,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus
gseaKEGG_MG_FR <- gseKEGG(
  geneList = Sort_log2fc_MG_FR,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus
gseaKEGG_MG_MR <- gseKEGG(
  geneList = Sort_log2fc_MG_MR,
  organism = "mmu",
  pvalueCutoff = 1,
  pAdjustMethod = "none"
) # Mus Musculus


View(gseaKEGG_AC_FM@result)
View(gseaKEGG_AC_FR@result)
View(gseaKEGG_AC_MR@result)

View(gseaKEGG_MG_FM@result)
View(gseaKEGG_MG_FR@result)
View(gseaKEGG_MG_MR@result)

# Plot the GSEA plot for one of the enriched pathways observed in the enrichedKEGG object
# Gen Set ID mmu00010
# AC
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05200", title = "GSEA enrichment score plot Astrocytes Front & Middle")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu01100", title = "GSEA enrichment score plot Astrocytes Front & Middle")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05010", title = "GSEA enrichment score plot Astrocytes Front & Middle")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05208", title = "GSEA enrichment score plot Astrocytes Front & Middle")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu04080", title = "GSEA enrichment score plot Astrocytes Front & Middle")
gseaplot(gseaKEGG_AC_FM, geneSetID = "mmu05022", title = "GSEA enrichment score plot Astrocytes Front & Middle")


gseaplot(gseaKEGG_AC_FR, geneSetID = "mmu01100", title = "GSEA enrichment score plot Astrocytes Front & Rare")
gseaplot(gseaKEGG_AC_FR, geneSetID = "mmu00480", title = "GSEA enrichment score plot Astrocytes Front & Rare")
gseaplot(gseaKEGG_AC_FR, geneSetID = "mmu05418", title = "GSEA enrichment score plot Astrocytes Front & Rare")
gseaplot(gseaKEGG_AC_FR, geneSetID = "mmu05207", title = "GSEA enrichment score plot Astrocytes Front & Rare")
gseaplot(gseaKEGG_AC_FR, geneSetID = "mmu05034", title = "GSEA enrichment score plot Astrocytes Front & Rare")

gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu04919", title = "GSEA enrichment score plot Astrocytes Middle & Rare")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu01100", title = "GSEA enrichment score plot Astrocytes Middle & Rare")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu04217", title = "GSEA enrichment score plot Astrocytes Middle & Rare")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu04530", title = "GSEA enrichment score plot Astrocytes Middle & Rare")
gseaplot(gseaKEGG_AC_MR, geneSetID = "mmu01212", title = "GSEA enrichment score plot Astrocytes Middle & Rare")

# MG
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu05200", title = "GSEA enrichment score plot Microglia Front & Middle")
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu05168", title = "GSEA enrichment score plot Microglia Front & Middle")
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu01100", title = "GSEA enrichment score plot Microglia Front & Middle")
gseaplot(gseaKEGG_MG_FM, geneSetID = "mmu04141", title = "GSEA enrichment score plot Microglia Front & Middle")


gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu05020", title = "GSEA enrichment score plot Microglia Front & Rare")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu01100", title = "GSEA enrichment score plot Microglia Front & Rare")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu04714", title = "GSEA enrichment score plot Microglia Front & Rare")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu00190", title = "GSEA enrichment score plot Microglia Front & Rare")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu05203", title = "GSEA enrichment score plot Microglia Front & Rare")
gseaplot(gseaKEGG_MG_FR, geneSetID = "mmu04142", title = "GSEA enrichment score plot Microglia Front & Rare")


gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu04810", title = "GSEA enrichment score plot Microglia Middle & Rare")
gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu05165", title = "GSEA enrichment score plot Microglia Middle & Rare")
gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu04510", title = "GSEA enrichment score plot Microglia Middle & Rare")
gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu05203", title = "GSEA enrichment score plot Microglia Middle & Rare")
gseaplot(gseaKEGG_MG_MR, geneSetID = "mmu05167", title = "GSEA enrichment score plot Microglia Middle & Rare")

```

```{r}
# read the data file with all samples
rawCounts <-
  read.table("raw_counts_cpm_after_norm_reorganised_220722.txt", header = TRUE)


# Load sample information data
sampleInfo <- read.table("Sample-info_220722.txt", header = TRUE)

# Load cell type marker genes data from an Excel file
cellType_markerGenes <- read_excel("Celltype_markergenes.xlsx")



ensembl <- useMart("ensembl")

listDatasets(ensembl)

ensembl_mouse <-
  useDataset("mmusculus_gene_ensembl", mart = ensembl)

mouse_gene_ids <- rawCounts$ensembl_id
mouse_gene_info <-
  getBM(
    attributes = c(
      "ensembl_gene_id",
      "external_gene_name",
      "chromosome_name",
      "start_position",
      "end_position",
      "gene_biotype",
      "strand"
    ),
    filters = "ensembl_gene_id",
    values = mouse_gene_ids,
    mart = ensembl_mouse
  )

# Changing name of column too merge the values with a matching column.
colnames(mouse_gene_info)[c(1)] <- c("ensembl_id")
# Merging the two dataframes into one dataframe leaving out one
rawCounts <- merge(rawCounts, mouse_gene_info, c(1))

rawCounts <- rawCounts[, -c(1, 38:42)]

colnames(rawCounts)[c(36)] <-  c("GeneNames")

# keep only the first occurrence of each duplicated gene name
rawCounts <- rawCounts[!duplicated(rawCounts$GeneNames), ]

# add gene name to  row names
row.names(rawCounts) <-  rawCounts$GeneNames
rawCounts <- rawCounts[, -36]

# Filter the data using logical indexing
# rawCounts <- rawCounts[rowSums(rawCounts >= 40) > 0, ]

# separating the columns if ribotag and totalRNA
# into  separated dataframe object for the rawCounts and sampleInfo
rawCounts_RT <- rawCounts[, grep("_RT", colnames(rawCounts))]
rawCounts_TR <- rawCounts[, grep("_TR", colnames(rawCounts))]

sampleInfo_RT <- sampleInfo[grep("_RT", sampleInfo$sample), ]
sampleInfo_TR <- sampleInfo[grep("_TR", sampleInfo$sample), ]


rawCounts_RT <- rawCounts_RT[order(colnames(rawCounts_RT))]
sampleInfo_RT <- sampleInfo_RT[order(sampleInfo_RT$sample),]

rawCounts_TR <- rawCounts_TR[order(colnames(rawCounts_TR))]
sampleInfo_TR <- sampleInfo_TR[order(sampleInfo_TR$sample),]

# Log the counts
logCounts_RT <- log2(rawCounts_RT + 1)
logCounts_TR <- log2(rawCounts_TR + 1)


# Adding marker genes before z-score
dfCounts_RT <-
  na.omit(logCounts_RT[cellType_markerGenes$gene_symbol,])
dfCounts_TR <-
  na.omit(logCounts_TR[cellType_markerGenes$gene_symbol,])

# making function for calculating z score since scale function is not ideal
calc_zscore <- function(df) {
  result <- (df - mean(df)) / sd(df)
  return(result)
}


# calculating the z-score
Z_dfCounts_RT <- apply(dfCounts_RT, 1, calc_zscore)
Z_dfCounts_TR <- apply(dfCounts_TR, 1, calc_zscore)

# combining the Ribotog z-score and totalRNA z-score and creating a factor for the sample names
sampleS <-
  factor(c(row.names(Z_dfCounts_RT), row.names(Z_dfCounts_TR)))
Combined_RT_TR <-
  merge(Z_dfCounts_RT,
        Z_dfCounts_TR,
        by = colnames(Z_dfCounts_TR),
        all = T)

# add a new column to each matrix with the factor as row names
row.names(Combined_RT_TR) <- sampleS
Combined_RT_TR <- data.matrix(Combined_RT_TR)


# Create a new data frame with gene_symbol and cell_type for row annotation
row_annotation <- data.frame(CellType = cellType_markerGenes$cell_type,
                             row.names = cellType_markerGenes$gene_symbol)

# Define a custom color scheme for cell types
cell_type_colors <- c("blue", "red", "green", "purple", "orange")
names(cell_type_colors) <-
  c("astro", "GABA", "Glut", "micro", "oligo")

# Generate the custom annotation dataframe
annotation <- data.frame(row_annotation)
rownames(annotation) <- rownames(row_annotation)
annotation_colors <- list(CellType = cell_type_colors)

library(pheatmap)
# Generate the heatmaps with the marker genes clustered by each cell type

pheatmap(
  t(Z_dfCounts_RT),
  color = colorRampPalette(c("purple", "white", "yellow"))(40),
  cluster_cols = T,
  cluster_rows = F,
  annotation_row = annotation,
  annotation_colors = annotation_colors,
  main = "Z-score of marker genes across cells Microglia & Astrocytes, condition Ribotag",
  fontsize = 8,
  cellwidth = 12,
  cellheight = 7
)

pheatmap(
  t(Z_dfCounts_TR),
  color = colorRampPalette(c("#CC99FF", "#FFFFFF", "#FFFF99"))(40),
  cluster_cols = T,
  cluster_rows = F,
  annotation_row = annotation,
  annotation_colors = annotation_colors,
  main = "Z-score of marker genes across cells Microglia & Astrocytes, condition totalRNA",
  fontsize = 8,
  cellwidth = 12,
  cellheight = 7
)

```

```{r}
# read the data file with all samples
rawCounts <-
  read.table("raw_counts_cpm_after_norm_reorganised_220722.txt", header = TRUE)


# Load sample information data
sampleInfo <- read.table("Sample-info_220722.txt", header = TRUE)

# Load cell type marker genes data from an Excel file
ribo_prot <- read_xlsx("ribo_prot.xlsx")

ensembl <- useMart("ensembl")

listDatasets(ensembl)

ensembl_mouse <-
  useDataset("mmusculus_gene_ensembl", mart = ensembl)

mouse_gene_ids <- rawCounts$ensembl_id
mouse_gene_info <-
  getBM(
    attributes = c(
      "ensembl_gene_id",
      "external_gene_name",
      "chromosome_name",
      "start_position",
      "end_position",
      "gene_biotype",
      "strand"
    ),
    filters = "ensembl_gene_id",
    values = mouse_gene_ids,
    mart = ensembl_mouse
  )

# Changing name of column too merge the values with a matching column.
colnames(mouse_gene_info)[c(1)] <- c("ensembl_id")
# Merging the two dataframes into one dataframe leaving out one
rawCounts <- merge(rawCounts, mouse_gene_info, c(1))

rawCounts <- rawCounts[, -c(1, 38:42)]

colnames(rawCounts)[c(36)] <-  c("GeneNames")

# keep only the first occurrence of each duplicated gene name
rawCounts <- rawCounts[!duplicated(rawCounts$GeneNames), ]

# add gene name to  row names
row.names(rawCounts) <-  rawCounts$GeneNames
rawCounts <- rawCounts[, -36]

# Filter the data using logical indexing
# rawCounts <- rawCounts[rowSums(rawCounts >= 40) > 0, ]

# separating the columns if ribotag and totalRNA
# into  separated dataframe object for the rawCounts and sampleInfo
rawCounts_RT <- rawCounts[, grep("_RT", colnames(rawCounts))]
rawCounts_TR <- rawCounts[, grep("_TR", colnames(rawCounts))]

sampleInfo_RT <- sampleInfo[grep("_RT", sampleInfo$sample), ]
sampleInfo_TR <- sampleInfo[grep("_TR", sampleInfo$sample), ]


rawCounts_RT <- rawCounts_RT[order(colnames(rawCounts_RT))]
sampleInfo_RT <- sampleInfo_RT[order(sampleInfo_RT$sample),]

rawCounts_TR <- rawCounts_TR[order(colnames(rawCounts_TR))]
sampleInfo_TR <- sampleInfo_TR[order(sampleInfo_TR$sample),]

# Log the counts
logCounts_RT <- log2(rawCounts_RT + 1)
logCounts_TR <- log2(rawCounts_TR + 1)


# Adding marker genes before z-score
dfCounts_RT_D <- na.omit(logCounts_RT[ribo_prot$Symbols,])
dfCounts_TR_D <- na.omit(logCounts_TR[ribo_prot$Symbols,])

# making function for calculating z score since scale function is not ideal
calc_zscore <- function(df) {
  result <- (df - mean(df)) / sd(df)
  return(result)
}


# calculating the z-score
Z_dfCounts_RT_D <- apply(dfCounts_RT_D, 1, calc_zscore)
Z_dfCounts_TR_D <- apply(dfCounts_TR_D, 1, calc_zscore)

# combining the Ribotog z-score and totalRNA z-score and creating a factor for the sample names
sampleS_D <-
  factor(c(row.names(Z_dfCounts_RT_D), row.names(Z_dfCounts_TR_D)))
Combined_RT_TR_D <-
  merge(Z_dfCounts_RT_D,
        Z_dfCounts_TR_D,
        by = colnames(Z_dfCounts_TR_D),
        all = T)

# add a new column to each matrix with the factor as row names
row.names(Combined_RT_TR_D) <- sampleS_D
Combined_RT_TR_D <- data.matrix(Combined_RT_TR_D)


# keep only the first occurrence of each duplicated gene name
ribo_prot <- ribo_prot[!duplicated(ribo_prot$Symbols), ]
# Create a new data frame with gene_symbol and cell_type for row annotation
row_annotation_D <- data.frame(Cromosome = ribo_prot$`Chromosome Number`,
                               row.names = ribo_prot$Symbols)

# Define a custom color scheme for cell types

# Generate the custom annotation dataframe
annotation_D <- data.frame(row_annotation_D)
rownames(annotation_D) <- rownames(row_annotation_D)

library(pheatmap)
# Generate the heatmaps with the marker genes clustered by each cell type
pheatmap(
  t(Z_dfCounts_RT_D),
  color = colorRampPalette(c("purple", "white", "yellow"))(40),
  cluster_cols = T,
  cluster_rows = F,
  main = "Z-score of Ribosomal protein encoded genes across cells Microglia & Astrocytes, condition Ribotag",
  fontsize = 8,
  cellwidth = 12,
  cellheight = 7
)

pheatmap(
  t(Z_dfCounts_TR_D),
  color = colorRampPalette(c("#CC99FF", "#FFFFFF", "#FFFF99"))(40),
  cluster_cols = T,
  cluster_rows = F,
  main = "Z-score of Ribosomal protein encoded genes across cells Microglia & Astrocytes, condition totalRNA",
  fontsize = 8,
  cellwidth = 12,
  cellheight = 7
)

```


```{r}



# Attaching data to object in the environment
# Sample Info
sampleInfo <- read.table("Sample-info_220722.txt", header = T)

# Celltype info
Celltype_markergenes <- read_excel("Celltype_markergenes.xlsx")

# Counts per million after normalization
rawCounts <-
  read.table(
    "raw_counts_cpm_after_norm_reorganised_220722.txt",
    header = T,
    row.names = 1
  )



# Set a seed to produce reproducible result
set.seed(16)

#Create DGE through DGEList() funtion
dgList <- DGEList(counts = rawCounts, genes = rownames(rawCounts))

# Add the replication column to the markergenes dataframe
# Building groups for the replication
replication <-
  factor(
    c(
      'rep1',
      'rep1',
      'rep1',
      'rep2',
      'rep2',
      'rep2',
      'rep3',
      'rep3',
      'rep2',
      'rep2',
      'rep2',
      'rep3',
      'rep3',
      'rep3',
      'rep1',
      'rep1',
      'rep1',
      'rep1',
      'rep1',
      'rep1',
      'rep2',
      'rep2',
      'rep2',
      'rep3',
      'rep3',
      'rep3',
      'rep2',
      'rep2',
      'rep2',
      'rep3',
      'rep3',
      'rep3',
      'rep1',
      'rep1',
      'rep1'
    )
  )

replication <- data.frame(replication)
sampleInfo$replication <- replication$replication



# Density plots of CPM per sample of data, filtered, and normalized log2
plotDensity(
  cpm(rawCounts, log = T),
  col = "red",
  xlab = "CPM of log2",
  ylab = "Density",
  main = "Normalized CPM log2 Data"
)

# Extract the Counts Matrix from the DGEList
Count_dgList <- dgList$counts

# Log the counts
logCounts_dgList <- log2(Count_dgList + 1)

# Filter out low-expressed genes
maxTPM <- apply(logCounts_dgList, 1, max)
minTPM <- apply(logCounts_dgList, 1, min)

logCounts_dgList <-
  logCounts_dgList[(maxTPM >= 2) & ((maxTPM - minTPM) >= 2),]

# Keep the top 500 genes based on variance
keepVariance <-
  order(apply(logCounts_dgList, 1, var), decreasing = T)[1:500]
topVariance_counts <- logCounts_dgList[keepVariance,]

# apply PCA on the transposed data
outputPCA <- prcomp(t(topVariance_counts))

# Calculate the variance of each PC
variancePC <- outputPCA$sdev ^ 2 / sum(outputPCA$sdev ^ 2)
variancePC <- round(variancePC, digits = 2)

variancePC_df <- data.frame(PC = paste0("PC", 1:10),
                            variancePC = variancePC[1:10])

# Create the scree plot
ggplot(variancePC_df, aes(x = PC, y = variancePC)) +
  geom_col() + geom_bar(stat = "identity", fill = "royalblue4") +
  labs(title = "Scree Plot of Principal Components") +
  ylab("Explained variance") + ylim(0, 1)

# Extract the PCs 1 and 2
outputPCA <- as.data.frame(outputPCA$x[, 1:2])

# Add DGEList varaibles to the PCA output
outputPCA[, 3:7] <- sampleInfo[, 1:5]
# Plot samples by their PC1 and PC2
ggplot(
  outputPCA,
  aes_string(
    x = colnames(outputPCA)[1],
    y = colnames(outputPCA)[2],
    color = "sample",
    shape = "replication"
  )
) +
  ggtitle("PCA Plot of Replication") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(outputPCA)[1], variancePC[1], sep = ": ")) +
  ylab(paste(colnames(outputPCA)[2], variancePC[2], sep = ": "))

ggplot(
  outputPCA,
  aes_string(
    x = colnames(outputPCA)[1],
    y = colnames(outputPCA)[2],
    color = "sample",
    shape = "Condition"
  )
) +
  ggtitle("PCA Plot of Condition") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(outputPCA)[1], variancePC[1], sep = ": ")) +
  ylab(paste(colnames(outputPCA)[2], variancePC[2], sep = ": "))



ggplot(outputPCA,
       aes_string(
         x = colnames(outputPCA)[1],
         y = colnames(outputPCA)[2],
         color = "sample",
         shape = "Region"
       )) +
  ggtitle("PCA Plot of Region") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(outputPCA)[1], variancePC[1], sep = ": ")) +
  ylab(paste(colnames(outputPCA)[2], variancePC[2], sep = ": "))

ggplot(
  outputPCA,
  aes_string(
    x = colnames(outputPCA)[1],
    y = colnames(outputPCA)[2],
    color = "sample",
    shape = "Cell_type"
  )
) +
  ggtitle("PCA Plot of Celltype") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(outputPCA)[1], variancePC[1], sep = ": ")) +
  ylab(paste(colnames(outputPCA)[2], variancePC[2], sep = ": "))







# Create a design matrix with our variable of interest
designMatrix <- model.matrix( ~ as.factor(Region), data = sampleInfo)

# Apply the Combat correction of the replicates
outputComBat <- ComBat(
  dat = topVariance_counts,
  batch = as.factor(sampleInfo$replication),
  mod = designMatrix,
  par.prior = F
)

# Set negative values to 0
outputComBat[outputComBat < 0] <- 0

# Apply again PCA
correctedPCA <- prcomp(t(outputComBat))


corrected_varExplained <-
  correctedPCA$sdev ^ 2 / sum(correctedPCA$sdev ^ 2)
corrected_varExplained <- round(corrected_varExplained, digits = 2)

correctedPCA <- as.data.frame(correctedPCA$x[, 1:2])

correctedPCA[, 3:7] <- sampleInfo[, 1:5]
# Plot PC1 and PC2 for the corrected data
ggplot(
  correctedPCA,
  aes_string(
    x = colnames(correctedPCA)[1],
    y = colnames(correctedPCA)[2],
    color = "sample",
    shape = "Region"
  )
) +
  ggtitle("PCA Plot of Replication Corrected Data") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(correctedPCA)[1], corrected_varExplained[2], sep = ": ")) +
  ylab(paste(colnames(correctedPCA)[2], corrected_varExplained[2], sep = ": "))


# Create a design matrix with our variable of interest
designMatrix1 <- model.matrix( ~ as.factor(Region), data = sampleInfo)

# Apply the Combat correction of the replicates
outputComBat1 <- ComBat(
  dat = topVariance_counts,
  batch = as.factor(sampleInfo$replication),
  mod = designMatrix,
  par.prior = F
)

# Set negative values to 0
outputComBat1[outputComBat1 < 0] <- 0

# Apply again PCA
correctedPCA1 <- prcomp(t(outputComBat1))


corrected_varExplained1 <-
  correctedPCA1$sdev ^ 2 / sum(correctedPCA1$sdev ^ 2)
corrected_varExplained1 <-
  round(corrected_varExplained1, digits = 2)

correctedPCA1 <- as.data.frame(correctedPCA1$x[, 1:2])

correctedPCA1[, 3:7] <- sampleInfo[, 1:5]

ggplot(
  correctedPCA1,
  aes_string(
    x = colnames(correctedPCA)[1],
    y = colnames(correctedPCA)[2],
    color = "sample",
    shape = "replication"
  )
) +
  ggtitle("PCA Plot of Replication Corrected Data") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(correctedPCA1)[1], corrected_varExplained1[2], sep = ": ")) +
  ylab(paste(colnames(correctedPCA1)[2], corrected_varExplained1[2], sep = ": "))

ggplot(
  correctedPCA1,
  aes_string(
    x = colnames(correctedPCA)[1],
    y = colnames(correctedPCA)[2],
    color = "sample",
    shape = "Condition"
  )
) +
  ggtitle("PCA Plot of Replication Corrected Data of Condition") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(correctedPCA1)[1], corrected_varExplained1[2], sep = ": ")) +
  ylab(paste(colnames(correctedPCA1)[2], corrected_varExplained1[2], sep = ": "))

ggplot(
  correctedPCA1,
  aes_string(
    x = colnames(correctedPCA)[1],
    y = colnames(correctedPCA)[2],
    color = "sample",
    shape = "Region"
  )
) +
  ggtitle("PCA Plot of Replication Corrected Data of Region") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(correctedPCA1)[1], corrected_varExplained1[2], sep = ": ")) +
  ylab(paste(colnames(correctedPCA1)[2], corrected_varExplained1[2], sep = ": "))

ggplot(
  correctedPCA1,
  aes_string(
    x = colnames(correctedPCA)[1],
    y = colnames(correctedPCA)[2],
    color = "sample",
    shape = "Cell_type"
  )
) +
  ggtitle("PCA Plot of Replication Corrected Data of Celltype") +
  geom_point(size = 3, alpha = 1) +
  theme(text = element_text(size = 3)) +
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(colour = "black")
  ) +
  xlab(paste(colnames(correctedPCA1)[1], corrected_varExplained1[2], sep = ": ")) +
  ylab(paste(colnames(correctedPCA1)[2], corrected_varExplained1[2], sep = ": "))


```

```{r}



#
sign_log2FC_AC_FM <- sign_PvalueAC_FM[, 2:3]
rownames(sign_log2FC_AC_FM) <- sign_log2FC_AC_FM$GeneNames
sign_log2FC_AC_FM <- sign_log2FC_AC_FM[, -1]
#
sign_log2FC_AC_FR <- sign_PvalueAC_FR[, 2:3]
rownames(sign_log2FC_AC_FR) <- sign_log2FC_AC_FR$GeneNames
sign_log2FC_AC_FR <- sign_log2FC_AC_FR[, -1]
#
sign_log2FC_AC_MR <- sign_PvalueAC_MR[, 2:3]
rownames(sign_log2FC_AC_MR) <- sign_log2FC_AC_MR$GeneNames
sign_log2FC_AC_MR <- sign_log2FC_AC_MR[, -1]

# Ensure that 'sign_log2FC_AC_FM' is a data frame
sign_log2FC_AC_FM <- as.data.frame(sign_log2FC_AC_FM)
sign_log2FC_AC_FR <- as.data.frame(sign_log2FC_AC_FR)
sign_log2FC_AC_MR <- as.data.frame(sign_log2FC_AC_MR)



#
sign_log2FC_MG_FM <- sign_PvalueMG_FM[, 2:3]
rownames(sign_log2FC_MG_FM) <- sign_log2FC_MG_FM$GeneNames
sign_log2FC_MG_FM <- sign_log2FC_MG_FM[, -1]
#
sign_log2FC_MG_FR <- sign_PvalueMG_FR[, 2:3]
rownames(sign_log2FC_MG_FR) <- sign_log2FC_MG_FR$GeneNames
sign_log2FC_MG_FR <- sign_log2FC_MG_FR[, -1]
#
sign_log2FC_MG_MR <- sign_PvalueMG_MR[, 2:3]
rownames(sign_log2FC_MG_MR) <- sign_log2FC_MG_MR$GeneNames
sign_log2FC_MG_MR <- sign_log2FC_MG_MR[, -1]

# Ensure that 'sign_log2FC_MG_FM' is a data frame
sign_log2FC_MG_FM <- as.data.frame(sign_log2FC_MG_FM)
sign_log2FC_MG_FR <- as.data.frame(sign_log2FC_MG_FR)
sign_log2FC_MG_MR <- as.data.frame(sign_log2FC_MG_MR)


# Now, 'sign_log2FC_MG_FM', 'sign_log2FC_MG_FR', 'sign_log2FC_MG_MR'
# should be a character vector of row names
sign_log2FC_MG_FM <- rownames(sign_log2FC_MG_FM)
sign_log2FC_MG_FR <- rownames(sign_log2FC_MG_FR)
sign_log2FC_MG_MR <- rownames(sign_log2FC_MG_MR)

head(sign_log2FC_MG_FM)
head(sign_log2FC_MG_FR)
head(sign_log2FC_MG_MR)



################### GENE SETS #############

# Create gene sets
gene_set_AC_FM <- as.character(sign_PvalueAC_FM$GeneNames)
gene_set_AC_FR <- as.character(sign_PvalueAC_FR$GeneNames)
gene_set_AC_MR <- as.character(sign_PvalueAC_MR$GeneNames)

gene_set_MG_FM <- as.character(sign_PvalueMG_FM$GeneNames)
gene_set_MG_FR <- as.character(sign_PvalueMG_FR$GeneNames)
gene_set_MG_MR <- as.character(sign_PvalueMG_MR$GeneNames)


# AC_FMR
gene_sets_AC_FMR <-
  list(
    Front_vs_Middle = gene_set_AC_FM,
    Front_vs_Rear = gene_set_AC_FR,
    Middle_vs_Rear = gene_set_AC_MR
  )
plot(
  euler(gene_sets_AC_FMR),
  quantities = TRUE,
  fills = c("red", "green", "yellow")
)

grid.text(
  "Venn Diagram of DGE's for Astrocytes across Front, Middle, and Rear Regions",
  x = 0.5,
  y = 0.85
)
# Find common genes
common_genes_AC <-
  Reduce(intersect,
         list(gene_set_AC_FM, gene_set_AC_FR, gene_set_AC_MR))

# Print common genes
print(common_genes_AC)

# Save common_genes to a file
write.csv(common_genes_AC, "common_genes_AC.csv")

# MG_FMR
gene_sets_MG_FMR <-
  list(
    Front_vs_Middle = gene_set_MG_FM,
    Front_vs_Rear = gene_set_MG_FR,
    Middle_vs_Rear = gene_set_MG_MR
  )
plot(
  euler(gene_sets_MG_FMR),
  quantities = TRUE,
  fills = c("red", "green", "yellow")
)
grid.text(
  "Venn Diagram of DGE's for Microglia across Front, Middle, and Rear Regions",
  x = 0.5,
  y = 0.85
)
# Find common genes
common_genes_MG <-
  Reduce(intersect,
         list(gene_set_MG_FM, gene_set_MG_FR, gene_set_MG_MR))

# Print common genes
print(common_genes_MG)

# Save common_genes to a file
write.csv(common_genes_MG, "common_genes_MG.csv")

## FM
# Create a named list of gene sets
gene_sets_AC_MG_FM <-
  list(Astrocyte_Front_vs_Middle = gene_set_AC_FM,
       Microglia_Front_vs_Middle = gene_set_MG_FM)

# Create an Euler diagram (a type of Venn diagram) with custom plot parameters
plot(
  euler(gene_sets_AC_MG_FM),
  quantities = TRUE,
  fills = c("red", "green")
)
grid.text(
  "Venn Diagram of DGE's for Astrocytes and Microglia Front vs Middle",
  x = 0.5,
  y = 0.80
)
# Find common genes
common_genes_MG_AC_FM <-
  Reduce(intersect, list(gene_set_MG_FM, gene_set_AC_FM))

# Print common genes
print(common_genes_MG_AC_FM)


## FR
# Create a named list of gene sets
gene_sets_AC_MG_FR <-
  list(Astrocyte_Front_vs_Rear = gene_set_AC_FR,
       Microglia_Front_vs_Rear = gene_set_MG_FR)

# Create an Euler diagram (a type of Venn diagram) with custom plot parameters
plot(
  euler(gene_sets_AC_MG_FR),
  quantities = TRUE,
  fills = c("red", "green")
)
grid.text(
  "Venn Diagram of DGE's for Astrocytes and Microglia Front vs Rear",
  x = 0.5,
  y = 0.83
)

# Find common genes
common_genes_MG_AC_FR <-
  Reduce(intersect, list(gene_set_MG_FR, gene_set_AC_FR))

# Print common genes
print(common_genes_MG_AC_FR)

## MR
# Create a named list of gene sets
gene_sets_AC_MG_MR <-
  list(Astrocyte_Middle_vs_Rear = gene_set_AC_MR,
       Microglia_Middle_vs_Rear = gene_set_MG_MR)

# Create an Euler diagram (a type of Venn diagram) with custom plot parameters
plot(
  euler(gene_sets_AC_MG_MR),
  quantities = TRUE,
  fills = c("red", "green")
)
grid.text(
  "Venn Diagram of DGE's for Astrocytes and Microglia Middle vs Rear",
  x = 0.5,
  y = 0.85
)
# Find common genes
common_genes_MG_AC_MR <-
  Reduce(intersect, list(gene_set_MG_MR, gene_set_AC_MR))

# Print common genes
print(common_genes_MG_AC_MR)

# Save all the intersected genes
write.csv(common_genes_MG_AC_FM, "common_genes_MG_AC_FM.csv")
write.csv(common_genes_MG_AC_FR, "common_genes_MG_AC_FR.csv")
write.csv(common_genes_MG_AC_MR, "common_genes_MG_AC_MR.csv")

```

```{r}
# separating the columns if ribotag and totalRNA  
# into  separated dataframe object for the rawCounts and sampleInfo
rawCounts_RT <- rawCounts[, grep("_RT", colnames(rawCounts))]
rawCounts_TR <- rawCounts[, grep("_TR", colnames(rawCounts))]

rawCounts_RT <- rawCounts_RT[order(colnames(rawCounts_RT))]

rawCounts_TR <- rawCounts_TR[order(colnames(rawCounts_TR))]


# making function for calculating z score since scale function is not ideal
calc_zscore <- function(df) {
  result <- (df - mean(df)) / sd(df)
  return(result)
}

# calculating the Z-score of ribotag over the Z-score of totalRNA
# calculating the z-score
Z_score_RT <- apply(rawCounts_RT, 1, calc_zscore)
Z_score_TR <- apply(rawCounts_TR, 1, calc_zscore)

# removing the 3rd sample from the TR list since it has no counterpart from ribotag data
Z_score_TR <- Z_score_TR[-3, ]

# Making a function for the Ribotag over totalRNA ratio
divide_rows <- function(x, y) {
  result <-
    matrix(
      nrow = nrow(x),
      ncol = ncol(x),
      dimnames = list(row.names(x), colnames(x))
    )
  for (i in 1:nrow(x)) {
    RT_sample_name <- row.names(x)[i]
    TR_sample_name <- gsub("RT", "TR", RT_sample_name)
    result[i,] <- x[i,] / y[TR_sample_name,]
    row.names(result)[i] <- gsub("RT", "RT_TR", RT_sample_name)
  }
  return(result)
}

RT_over_TR <- t(divide_rows(Z_score_RT, Z_score_TR))

#filtering intervall
RT_over_TR <- RT_over_TR[apply(RT_over_TR, 1, function(x) {all(x > -2 & x < 2)}),]


sample_ratios <- list(
  AC_F = RT_over_TR[, grep("AC_F", colnames(RT_over_TR))],
  AC_M = RT_over_TR[, grep("AC_M", colnames(RT_over_TR))],
  AC_R = RT_over_TR[, grep("AC_R", colnames(RT_over_TR))],
  MG_F = RT_over_TR[, grep("MG_F", colnames(RT_over_TR))],
  MG_M = RT_over_TR[, grep("MG_M", colnames(RT_over_TR))],
  MG_R = RT_over_TR[, grep("MG_R", colnames(RT_over_TR))]
)

# Extract the gene sets for Astrocytes and Microglia in all regions from the sample_ratio list
genes_list <- list(AC_F = sample_ratios$AC_F,
                   AC_M = sample_ratios$AC_M,
                   AC_R = sample_ratios$AC_R,
                   MG_F = sample_ratios$MG_F,
                   MG_M = sample_ratios$MG_M,
                   MG_R = sample_ratios$MG_R)
# Convert the list to a binary incidence matrix
binary_matrix <- fromList(genes_list)

# Create the UpSet plot
upset(
  binary_matrix,
  sets = c("AC_F", "AC_M", "AC_R", "MG_F", "MG_M", "MG_R"),
  order.by = "freq",
  mainbar.y.label = "Intersection Size",
  sets.bar.color = "black"
)

```

